{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Moment dynamics with cell division"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "(c) 2017 Manuel Razo. This work is licensed under a [Creative Commons Attribution License CC-BY 4.0](https://creativecommons.org/licenses/by/4.0/). All code contained herein is licensed under an [MIT license](https://opensource.org/licenses/MIT)\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "import cloudpickle\n",
    "\n",
    "# Our numerical workhorses\n",
    "import numpy as np\n",
    "import scipy as sp\n",
    "import pandas as pd\n",
    "\n",
    "# Import matplotlib stuff for plotting\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.cm as cm\n",
    "import matplotlib as mpl\n",
    "\n",
    "# Seaborn, useful for graphics\n",
    "import seaborn as sns\n",
    "\n",
    "# Import the utils for this project\n",
    "import chann_cap_utils as chann_cap\n",
    "\n",
    "# Set PBoC plotting style\n",
    "chann_cap.set_plotting_style()\n",
    "\n",
    "# Magic function to make matplotlib inline; other style specs must come AFTER\n",
    "%matplotlib inline\n",
    "\n",
    "# This enables SVG graphics inline\n",
    "%config InlineBackend.figure_format = 'svg'\n",
    "\n",
    "tmpdir = '../../tmp/'\n",
    "figdir = '../../fig/moment_dynamics_numeric/'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### $\\LaTeX$ macros\n",
    "$\\newcommand{kpon}{k^{(p)}_{\\text{on}}}$\n",
    "$\\newcommand{kpoff}{k^{(p)}_{\\text{off}}}$\n",
    "$\\newcommand{kron}{k^{(r)}_{\\text{on}}}$\n",
    "$\\newcommand{kroff}{k^{(r)}_{\\text{off}}}$\n",
    "$\\newcommand{rm}{r _m}$\n",
    "$\\newcommand{gm}{\\gamma _m}$\n",
    "$\\newcommand{mm}{\\left\\langle m \\right\\rangle}$\n",
    "$\\newcommand{ee}[1]{\\left\\langle #1 \\right\\rangle}$\n",
    "$\\newcommand{bb}[1]{\\mathbf{#1}}$"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Distribution moment dynamics with cell division "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As first discussed by Jones et al. and then further expanded by [Peterson et al.](http://www.pnas.org/content/112/52/15886) the effect of having multiple gene copy numbers due to genome replication during the cell cycle has an important effect on gene expression noise. As the genome is replicated the cells spend part of their cell cycle with > 1 copy of the gene. The number of copies depends on the growth rate and on the gene position relative to the replication origin.\n",
    "\n",
    "For our experimental setup we know from Jones et al. that our cells spend 2/3 of the cell cycle with 2 copies of the reporter gene and 1/3 with 1 copy. We previously inferred the parameters $\\kpon$, $\\kpoff$, and $r_m$ assuming that at both stages the mRNA reached steady state with $r_m$ as production rate for 1 gene copy and $2 r_m$ for two copies. The objective of this notebook is to explore the dynamical consequences of these gene copy number variations at the level of the mRNA and protein distribution moments."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The first thing we need to establish are the dynamics for the mRNA. The cell cycle time $t_d \\approx 60$ min establishes the period of the cell cycle. For a time $0 < t_s < t_d$ cells have 1 copy of the gene (i.e. mRNA production rate $r_m$) and for the rest of the cycle the cells have 2 copies of the gene (i.e. mRNA production rate of $2 r_m$).\n",
    "\n",
    "Therefore for our simulations we will initialize the moments at the steady state values for the single promoter, run the simulation with those parameters for time 0 to $t_s$ and then change the parameters for the rest of the simulation until reaching time $t_d$."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Cell division"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "After cells reach time $t_d$ the content of the cells is assumed to divide as they divide too. To account for this division we will assume that the mean mRNA (and later on protein) count are halved. Given the dependence of higher moments on the first moment we will \"re-initialize\" the value of the other moments assuming the values that one would obtain if this halved mean mRNA count was the steady state value. \n",
    "\n",
    "In other words assume that the mean mRNA count at time $t_d$ is given by $\\ee{m(t_d)}$. After this time we reset the initial condition for the simulation with one promoter to be $\\ee{m(t=0)} = {\\ee{m(t_d)} \\over 2}$. By fixing the first moment values higher moments of the mRNA distirbution $\\ee{m^x}$ will be computed by finding what the steady state value would be given the value of the first moment. Note that this will not strictly give the same values as just computing the steady state value of the moments. If the initial condition $\\ee{m(t=0)}$ is not the same as steady state condition $\\ee{m}_{ss}$ the values that the other moments will also differ from the steady state values.\n",
    "\n",
    "The reason for this convoluted assumption is that dividing the first moment in half after cell division makes sense, but it is not clear that the same assumption applies for higher moments of the distribution. To gain intuition about this assume that the mRNA distribution is binomial with mean value $\\ee{m(t_d)} = N p$. Dividing this by half for the initial condition is easy since we would have $\\ee{m(t=0)} = {N p \\over 2}$. The second moment before the cell division would be of the form $\\ee{m^2(t_d)} = Np - Np^2 + N^2p^2$. Therefore dividing this value by two for the initial condition of the next cell cycle doesn't translate to the second moment one would expect for a binomial distribution with mean ${Np \\over 2}$."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Defining mRNA dynamics "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will now define functions to compute the first 3 mRNA distribution moments. The PDEs that define these moments are given by\n",
    "$$\n",
    "{d \\over dt} \\ee{\\bb{m}^0} = \\mathbf{K}_m \\left\\langle \\mathbf{m}^0 \\right\\rangle,\n",
    "$$\n",
    "$$\n",
    "{d \\over dt}\\ee{\\bb{m}^1} = \n",
    "\\left( \\mathbf{K}_m - \\mathbf{\\Gamma}_m  \\right)\\ee{\\bb{m}^1}\n",
    "             + \\mathbf{R}_m \\left\\langle \\mathbf{m}^0 \\right\\rangle,\n",
    "$$\n",
    "$$\n",
    "{d\\ee{\\bb{m}^2} \\over dt} = \\left( \\mathbf{K}^u_m - 2 \\mathbf{\\Gamma}^u_n \\right) \n",
    "             \\left\\langle \\mathbf{m}^2 \\right\\rangle\n",
    "             + \\left( 2 \\mathbf{R}^u_m + \\mathbf{\\Gamma}^u_m \\right)\n",
    "             \\left\\langle \\mathbf{m}^1 \\right\\rangle\n",
    "             + \\mathbf{R}^u_m\n",
    "             \\left\\langle \\mathbf{m}^0 \\right\\rangle,\n",
    "$$\n",
    "and\n",
    "$$\n",
    "{d\\ee{\\bb{m}^3} \\over dt} = \\left( \\mathbf{K}_m - 3 \\mathbf{\\Gamma}_n \\right) \n",
    "             \\left\\langle \\mathbf{m}^3 \\right\\rangle\n",
    "             + \\left( 3 \\mathbf{R}_m + 3 \\mathbf{\\Gamma}_m \\right)\n",
    "             \\left\\langle \\mathbf{m}^2 \\right\\rangle\n",
    "             + \\left( 3 \\mathbf{R}_m - \\mathbf{\\Gamma}_m \\right)\n",
    "             \\left\\langle \\mathbf{m}^1 \\right\\rangle\n",
    "             + \\mathbf{R}_m\n",
    "             \\left\\langle \\mathbf{m}^0 \\right\\rangle.\n",
    "$$"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "From these equations it is clear that the second and third moment depend on the value of the first (and zeroth) moment. For the first function `dmdt` the equations for the first and zeroth moment will be included in the numerical integration. For the second function `dmdt_init` the values for $\\left\\langle \\mathbf{m}^0 \\right\\rangle$ and $\\left\\langle \\mathbf{m}^1 \\right\\rangle$ are assumed to be fixed and the rest of the moments are computed accordingly."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [],
   "source": [
    "def dmdt(m, t, Kmat, Rmat, Gmat):\n",
    "    '''\n",
    "    function to integrate dm / dt using scipy.integrate.odeint\n",
    "    Parameters\n",
    "    ----------\n",
    "    m : array-like.\n",
    "        Array containing the zeroth, first, second and third moment \n",
    "        of the states of the promoter.\n",
    "        Unregulated\n",
    "        m[0] = m0_P (RNAP bound)\n",
    "        m[1] = m0_E (Empty promoter)\n",
    "        m[2] = m1_P (RNAP bound)\n",
    "        m[3] = m1_P (Empty promoter)\n",
    "        m[4] = m2_P (RNAP bound)\n",
    "        m[5] = m2_P (Empty promoter)\n",
    "        m[6] = m3_P (RNAP bound)\n",
    "        m[7] = m3_P (Empty promoter)\n",
    "        ---------\n",
    "        Regulated:\n",
    "        m[0] = m0_P (RNAP bound)\n",
    "        m[1] = m0_E (Empty promoter)\n",
    "        m[2] = m0_R (Repressor bound) \n",
    "        m[3] = m1_P (RNAP bound)\n",
    "        m[4] = m1_E (Empty promoter)\n",
    "        m[5] = m1_R (Repressor bound) \n",
    "        m[6] = m2_P (RNAP bound)\n",
    "        m[7] = m2_E (Empty promoter)\n",
    "        m[8] = m2_R (Repressor bound) \n",
    "        m[9] = m3_P (RNAP bound)\n",
    "        m[10] = m3_E (Empty promoter)\n",
    "        m[11] = m3_R (Repressor bound) \n",
    "    t : array-like.\n",
    "        Time array\n",
    "    Kmat : array-like.\n",
    "        Matrix containing the transition rates between the promoter states.\n",
    "    Rmat : array-like.\n",
    "        Matrix containing the production rate at each of the states.\n",
    "    Gmat : array-like.\n",
    "        Matrix containing the degradation rate at each of the states.   \n",
    "        \n",
    "    Returns\n",
    "    -------\n",
    "    numerical integration of the differential equations\n",
    "    '''\n",
    "    # Obtain the zeroth and first moment based on the size \n",
    "    # of the Kmat matrix\n",
    "    if Kmat.shape[0] == 2:\n",
    "        m0 = m[0:2]\n",
    "        m1 = m[2:4]\n",
    "        m2 = m[4:6]\n",
    "        m3 = m[6::]\n",
    "    elif Kmat.shape[0] == 3:\n",
    "        m0 = m[0:3]\n",
    "        m1= m[3:6]\n",
    "        m2= m[6:9]\n",
    "        m3= m[9::]\n",
    "    \n",
    "    # Compute the moment equations for the:\n",
    "    # Zeroth moment\n",
    "    dm0dt_eq = np.dot(Kmat, m0)\n",
    "    # First moment\n",
    "    dm1dt_eq = np.dot((Kmat - Gmat), m1) + np.dot(Rmat, m0)\n",
    "    # Second moment\n",
    "    dm2dt_eq = np.dot((Kmat - 2 * Gmat), m2) + np.dot((2 * Rmat + Gmat), m1) +\\\n",
    "                np.dot(Rmat, m0)\n",
    "    # Third moment\n",
    "    dm3dt_eq = np.dot((Kmat - 3 * Gmat), m3) +\\\n",
    "                np.dot((3 * Rmat + 3 * Gmat), m2) +\\\n",
    "                np.dot((3 * Rmat - Gmat), m1) +\\\n",
    "                np.dot(Rmat, m0)\n",
    "    \n",
    "    return np.append(dm0dt_eq, \n",
    "                     np.append(dm1dt_eq, \n",
    "                               np.append(dm2dt_eq, dm3dt_eq)))\n",
    "\n",
    "def dmdt_init(m, t, m_fix, Kmat, Rmat, Gmat):\n",
    "    '''\n",
    "    function to integrate dm / dt using scipy.integrate.odeint\n",
    "    This funciton assumes that the zeroth and first moment are fixed\n",
    "    and computes the corresponding second and third moments accordingly.\n",
    "    Parameters\n",
    "    ----------\n",
    "    m : array-like.\n",
    "        Array containing the second and third moment \n",
    "        of the states of the promoter.\n",
    "        Unregulated\n",
    "        m[0] = m2_P (RNAP bound)\n",
    "        m[1] = m2_E (Empty promoter)\n",
    "        m[2] = m3_P (RNAP bound)\n",
    "        m[3] = m3_E (Empty promoter)\n",
    "        ---------\n",
    "        Regulated: \n",
    "        m[0] = m2_P (RNAP bound)\n",
    "        m[1] = m2_E (Empty promoter)\n",
    "        m[2] = m2_R (Repressor bound) \n",
    "        m[3] = m3_P (RNAP bound)\n",
    "        m[4] = m3_E (Empty promoter)\n",
    "        m[5] = m3_R (Repressor bound)\n",
    "        \n",
    "    m_fix : array-like.\n",
    "        Array containing the zerot and first moment \n",
    "        of the states of the promoter that will remain fixed.\n",
    "        Unregulated\n",
    "        m_fix[0] = m0_P (RNAP bound)\n",
    "        m_fix[1] = m0_E (Empty promoter)\n",
    "        m_fix[2] = m1_P (RNAP bound)\n",
    "        m_fix[3] = m1_E (Empty promoter)\n",
    "    ---------\n",
    "    Regulated:\n",
    "        m_fix[0] = m0_P (RNAP bound)\n",
    "        m_fix[1] = m0_E (Empty promoter)\n",
    "        m_fix[2] = m0_R (Repressor bound) \n",
    "        m_fix[3] = m1_P (RNAP bound)\n",
    "        m_fix[4] = m1_E (Empty promoter)\n",
    "        m_fix[5] = m1_R (Repressor bound)\n",
    "        \n",
    "    t : array-like.\n",
    "        Time array\n",
    "    Kmat : array-like.\n",
    "        Matrix containing the transition rates between the promoter states.\n",
    "    Rmat : array-like.\n",
    "        Matrix containing the production rate at each of the states.\n",
    "    Gmat : array-like.\n",
    "        Matrix containing the degradation rate at each of the states.   \n",
    "        \n",
    "    Returns\n",
    "    -------\n",
    "    2nd and 3rd moment dynamics given fixed 0th and 1st moment\n",
    "    '''\n",
    "    # Obtain the zeroth and first moment based on the size \n",
    "    # of the Kmat matrix\n",
    "    if Kmat.shape[0] == 2:\n",
    "        m0 = m_fix[0:2]\n",
    "        m1 = m_fix[2::]\n",
    "        m2 = m[0:2]\n",
    "        m3 = m[2::]\n",
    "    elif Kmat.shape[0] == 3:\n",
    "        m0 = m_fix[0:3]\n",
    "        m1= m_fix[3::]\n",
    "        m2= m[0:3]\n",
    "        m3= m[3::]\n",
    "    \n",
    "    # Compute the moment equations for the:\n",
    "    # Second moment\n",
    "    dm2dt_eq = np.dot((Kmat - 2 * Gmat), m2) + np.dot((2 * Rmat + Gmat), m1) +\\\n",
    "                np.dot(Rmat, m0)\n",
    "    # Third moment\n",
    "    dm3dt_eq = np.dot((Kmat - 3 * Gmat), m3) +\\\n",
    "                np.dot((3 * Rmat + 3 * Gmat), m2) +\\\n",
    "                np.dot((3 * Rmat - Gmat), m1) +\\\n",
    "                np.dot(Rmat, m0)\n",
    "    \n",
    "    return np.append(dm2dt_eq, dm3dt_eq)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Defining function to compute moments over severall cell cycles"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Having defined the functions to integrate the moment ODEs let's now define a function that given a time `t_single` that cells spend with one promoter copy and a time `t_double` that cells spend with two copies integrates the moments over several cell cycles."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [],
   "source": [
    "def dmdt_cycles(m_init, t_single, t_double, n_cycles, \n",
    "                Kmat, Rmat, Gmat, n_steps=300, t_ss=6000,\n",
    "                states=['P', 'E']):\n",
    "    '''\n",
    "    Function that integrates the mRNA moments over several cell cycles.\n",
    "    Parameters\n",
    "    ----------\n",
    "    m : array-like.\n",
    "        Array containing the zeroth, first, second and third moment \n",
    "        of the states of the promoter.\n",
    "        Unregulated\n",
    "        m[0] = m0_P (RNAP bound)\n",
    "        m[1] = m0_E (Empty promoter)\n",
    "        m[2] = m1_P (RNAP bound)\n",
    "        m[3] = m1_E (Empty promoter)\n",
    "        m[4] = m2_P (RNAP bound)\n",
    "        m[5] = m2_E (Empty promoter)\n",
    "        m[6] = m3_P (RNAP bound)\n",
    "        m[7] = m3_E (Empty promoter)\n",
    "        ---------\n",
    "        Regulated:\n",
    "        m[0] = m0_P (RNAP bound)\n",
    "        m[1] = m0_E (Empty promoter)\n",
    "        m[2] = m0_R (Repressor bound) \n",
    "        m[3] = m1_P (RNAP bound)\n",
    "        m[4] = m1_E (Empty promoter)\n",
    "        m[5] = m1_R (Repressor bound) \n",
    "        m[6] = m2_P (RNAP bound)\n",
    "        m[7] = m2_E (Empty promoter)\n",
    "        m[8] = m2_R (Repressor bound) \n",
    "        m[9] = m3_P (RNAP bound)\n",
    "        m[10] = m3_E (Empty promoter)\n",
    "        m[11] = m3_R (Repressor bound)\n",
    "    t_single : float.\n",
    "        Time [in 1/mRNA degradation rate units] that cells spend \n",
    "        with a single promoter copy\n",
    "    t_double : float.\n",
    "        Time [in 1/mRNA degradation rate units] that cells spend \n",
    "        with a two promoter copies.\n",
    "    n_cycles : int.\n",
    "        Number of cell cycles to integrate for. A cell cycle is defined\n",
    "        as t_single + t_double.\n",
    "    Kmat : array-like.\n",
    "        Matrix containing the transition rates between the promoter states.\n",
    "    Rmat : array-like.\n",
    "        Matrix containing the production rate at each of the states.\n",
    "    Gmat : array-like.\n",
    "        Matrix containing the degradation rate at each of the states.\n",
    "    n_steps : int. Default = 300.\n",
    "        Number of steps to use for the numerical integration.\n",
    "    t_ss : float.\n",
    "        Time to integrate for after cell division to compute the steady\n",
    "        state value of higher moments.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    mRNA moment dynamics over cell cycles\n",
    "    '''\n",
    "    # Initialize names for moments in data frame\n",
    "    names = ['m' + str(mom) + '_' + st for mom in range(4) for st in states]\n",
    "    \n",
    "    # Initialize data frame\n",
    "    df = pd.DataFrame(columns=['time', 'state', 'cycle'] + names)\n",
    "    \n",
    "    # Initilaize global time\n",
    "    t_sim = 0\n",
    "    # Loop through cycles\n",
    "    for cyc in range(n_cycles):\n",
    "        # == Single promoter == #\n",
    "        # Define time array\n",
    "        t = np.linspace(0, t_single, n_steps)\n",
    "\n",
    "        # Integrate moment equations\n",
    "        m = sp.integrate.odeint(dmdt, m_init, t, \n",
    "                             args=(Kmat, Rmat, Gmat))\n",
    "\n",
    "        # Generate data frame\n",
    "        df_m = pd.DataFrame(m, columns=names)\n",
    "        # Append time, state and cycle\n",
    "        df_m = df_m.assign(time=t + t_sim)\n",
    "        df_m = df_m.assign(state=['single'] * m.shape[0])\n",
    "        df_m = df_m.assign(cycle=[cyc] * m.shape[0])\n",
    "        \n",
    "        # Append results to global data frame\n",
    "        df = df.append(df_m, ignore_index=True, sort=False)\n",
    "        \n",
    "        # Update global time\n",
    "        t_sim = t_sim + t[-1]\n",
    "        \n",
    "        # == Two promoters == #\n",
    "        \n",
    "        # Define initial conditions as last point of single promoter state\n",
    "        m_init = m[-1, :]\n",
    "        \n",
    "        # Define time array\n",
    "        t = np.linspace(0, t_double, n_steps)\n",
    "\n",
    "        # Integrate moment equations\n",
    "        m = sp.integrate.odeint(dmdt, m_init, t, \n",
    "                             args=(Kmat, 2 * Rmat, Gmat))\n",
    "\n",
    "        # Generate data frame\n",
    "        df_m = pd.DataFrame(m, columns=names)\n",
    "        # Append time, state and cycle\n",
    "        df_m = df_m.assign(time=t + t_sim)\n",
    "        df_m = df_m.assign(state=['double'] * m.shape[0])\n",
    "        df_m = df_m.assign(cycle=[cyc] * m.shape[0])\n",
    "        \n",
    "        # Append results to global data frame\n",
    "        df = df.append(df_m, ignore_index=True, sort=False)\n",
    "        \n",
    "        # Update global time\n",
    "        t_sim = t_sim + t[-1]\n",
    "        \n",
    "        # == Cell division == #\n",
    "        \n",
    "        # Define time array\n",
    "        t = np.linspace(0, t_ss, n_steps)\n",
    "        \n",
    "        # Define initial conditions as last point of single promoter state\n",
    "        m_fix = m[-1, 0:(Kmat.shape[0]*2)]\n",
    "        m_init = m[-1, (Kmat.shape[0]*2)::]\n",
    "        \n",
    "        # Define initial conditions for moments after cell division\n",
    "        m = sp.integrate.odeint(dmdt_init, m_init, t, \n",
    "                             args=(m_fix, Kmat, Rmat, Gmat))\n",
    "        \n",
    "        m_init = np.append(m_fix, m[-1, :])\n",
    "        \n",
    "    return df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Two-state promoter\n",
    "Having defined these functions let's first test them with the two-state unregulated promoter.\n",
    "\n",
    "Let's define the necessary parameters."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define mRNA rate\n",
    "# http://bionumbers.hms.harvard.edu/bionumber.aspx?id=105717&ver=3&trm=lacZ%20mRNA%20lifetime&org=\n",
    "gm = 1 / (3 * 60)\n",
    "\n",
    "# Load the flat-chain\n",
    "with open('../../data/mcmc/lacUV5_constitutive_mRNA_double.pkl', 'rb') as file:\n",
    "    unpickler = pickle.Unpickler(file)\n",
    "    gauss_flatchain = unpickler.load()\n",
    "    gauss_flatlnprobability = unpickler.load()\n",
    "\n",
    "# Generate a Pandas Data Frame with the mcmc chain\n",
    "index = ['kp_on', 'kp_off', 'rm']\n",
    "\n",
    "# Generate a data frame out of the MCMC chains\n",
    "df_mcmc = pd.DataFrame(gauss_flatchain, columns=index)\n",
    "\n",
    "# rerbsine the index with the new entries\n",
    "index = df_mcmc.columns\n",
    "\n",
    "# map value of the parameters\n",
    "max_idx = np.argmax(gauss_flatlnprobability, axis=0)\n",
    "kp_on, kp_off, rm = df_mcmc.iloc[max_idx, :] * gm"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we will define the state transition matrix $\\mathbf{K}$, the mRNA production matrix $\\mathbf{R}_m$ and the mRNA degradation matrix $\\mathbf{\\Gamma}_m$."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the rate constant matrix\n",
    "Kmat_unreg = np.array([[-kp_off, kp_on], \n",
    "                       [kp_off, -kp_on]])\n",
    "# Define the production matrix\n",
    "Rmat_unreg = np.array([[rm, 0], \n",
    "                       [0, 0]])\n",
    "# Define the degradation matrix\n",
    "Gmat_unreg = np.array([[gm, 0],\n",
    "                       [0, gm]])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Having defined this let's integrate the equations for a long time to obtain the initial condition."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define time on which to perform integration\n",
    "t = np.linspace(0, 100 * 60, 301)\n",
    "\n",
    "# Define initial conditions\n",
    "m0_init = [0.5, 0.5]\n",
    "m1_init = [0, 0]\n",
    "m2_init = [0, 0]\n",
    "m3_init = [0, 0]\n",
    "\n",
    "# Numerically integrate equations\n",
    "m_init = sp.integrate.odeint(dmdt, m0_init + m1_init + m2_init + m3_init, t, \n",
    "                             args=(Kmat_unreg, Rmat_unreg, Gmat_unreg))\n",
    "\n",
    "m_init = m_init[-1, :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [],
   "source": [
    "t_single = 60 * 1 / 3 * 60 # sec\n",
    "t_double = 60 * 2 / 3 * 60 # sec\n",
    "n_cycles = 5\n",
    "\n",
    "test = dmdt_cycles(m_init, t_single, t_double, n_cycles, \n",
    "            Kmat_unreg, Rmat_unreg, Gmat_unreg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.text.Text at 0x1a1c120278>"
      ]
     },
     "execution_count": 132,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"utf-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       "  \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Created with matplotlib (http://matplotlib.org/) -->\n",
       "<svg height=\"275pt\" version=\"1.1\" viewBox=\"0 0 392 275\" width=\"392pt\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       " <defs>\n",
       "  <style type=\"text/css\">\n",
       "*{stroke-linecap:butt;stroke-linejoin:round;}\n",
       "  </style>\n",
       " </defs>\n",
       " <g id=\"figure_1\">\n",
       "  <g id=\"patch_1\">\n",
       "   <path d=\"M 0 275.836845 \n",
       "L 392.060937 275.836845 \n",
       "L 392.060937 0 \n",
       "L 0 0 \n",
       "z\n",
       "\" style=\"fill:none;\"/>\n",
       "  </g>\n",
       "  <g id=\"axes_1\">\n",
       "   <g id=\"patch_2\">\n",
       "    <path d=\"M 50.060938 229.688095 \n",
       "L 384.860938 229.688095 \n",
       "L 384.860938 12.248095 \n",
       "L 50.060938 12.248095 \n",
       "z\n",
       "\" style=\"fill:#e3dcd0;\"/>\n",
       "   </g>\n",
       "   <g id=\"matplotlib.axis_1\">\n",
       "    <g id=\"xtick_1\">\n",
       "     <g id=\"line2d_1\">\n",
       "      <path clip-path=\"url(#p9bb4136610)\" d=\"M 65.279119 229.688095 \n",
       "L 65.279119 12.248095 \n",
       "\" style=\"fill:none;stroke:#ffffff;stroke-dasharray:1.5,2.475;stroke-dashoffset:0;stroke-width:1.5;\"/>\n",
       "     </g>\n",
       "     <g id=\"line2d_2\"/>\n",
       "     <g id=\"text_1\">\n",
       "      <!-- 0 -->\n",
       "      <defs>\n",
       "       <path d=\"M 31.640625 -1.8125 \n",
       "Q 20.265625 -1.8125 13.296875 8.609375 \n",
       "Q 6.34375 19.046875 6.34375 36.140625 \n",
       "Q 6.34375 53.21875 13.296875 63.640625 \n",
       "Q 20.265625 74.078125 31.640625 74.078125 \n",
       "Q 43.0625 74.078125 50 63.671875 \n",
       "Q 56.9375 53.265625 56.9375 36.140625 \n",
       "Q 56.9375 19.046875 50 8.609375 \n",
       "Q 43.0625 -1.8125 31.640625 -1.8125 \n",
       "z\n",
       "M 31.640625 5.421875 \n",
       "Q 46.734375 5.421875 46.734375 36.140625 \n",
       "Q 46.734375 66.84375 31.640625 66.84375 \n",
       "Q 16.609375 66.84375 16.609375 36.140625 \n",
       "Q 16.609375 5.421875 31.640625 5.421875 \n",
       "z\n",
       "\" id=\"LucidaSansUnicode-30\"/>\n",
       "      </defs>\n",
       "      <g style=\"fill:#262626;\" transform=\"translate(61.168885 246.710283)scale(0.13 -0.13)\">\n",
       "       <use xlink:href=\"#LucidaSansUnicode-30\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"xtick_2\">\n",
       "     <g id=\"line2d_3\">\n",
       "      <path clip-path=\"url(#p9bb4136610)\" d=\"M 107.551847 229.688095 \n",
       "L 107.551847 12.248095 \n",
       "\" style=\"fill:none;stroke:#ffffff;stroke-dasharray:1.5,2.475;stroke-dashoffset:0;stroke-width:1.5;\"/>\n",
       "     </g>\n",
       "     <g id=\"line2d_4\"/>\n",
       "     <g id=\"text_2\">\n",
       "      <!-- 2500 -->\n",
       "      <defs>\n",
       "       <path d=\"M 8.296875 0 \n",
       "L 8.296875 8.453125 \n",
       "Q 12.5 18.265625 25.296875 29.828125 \n",
       "L 30.8125 34.765625 \n",
       "Q 41.453125 44.390625 41.453125 53.859375 \n",
       "Q 41.453125 59.90625 37.8125 63.375 \n",
       "Q 34.1875 66.84375 27.78125 66.84375 \n",
       "Q 20.3125 66.84375 10.109375 61.03125 \n",
       "L 10.109375 69.53125 \n",
       "Q 19.734375 74.078125 29.25 74.078125 \n",
       "Q 39.40625 74.078125 45.53125 68.609375 \n",
       "Q 51.65625 63.140625 51.65625 54.109375 \n",
       "Q 51.65625 47.609375 48.546875 42.578125 \n",
       "Q 45.453125 37.546875 37.015625 30.375 \n",
       "L 33.296875 27.203125 \n",
       "Q 21.734375 17.390625 19.921875 8.453125 \n",
       "L 51.3125 8.453125 \n",
       "L 51.3125 0 \n",
       "z\n",
       "\" id=\"LucidaSansUnicode-32\"/>\n",
       "       <path d=\"M 12.15625 -0.296875 \n",
       "L 12.15625 8.296875 \n",
       "Q 19.34375 5.421875 25.828125 5.421875 \n",
       "Q 32.953125 5.421875 37.03125 9.609375 \n",
       "Q 41.109375 13.8125 41.109375 21.1875 \n",
       "Q 41.109375 29.109375 35.34375 33.28125 \n",
       "Q 29.59375 37.453125 18.703125 37.453125 \n",
       "Q 16.15625 37.453125 13.375 37.109375 \n",
       "L 13.375 72.265625 \n",
       "L 49.953125 72.265625 \n",
       "L 49.953125 63.875 \n",
       "L 21.828125 63.875 \n",
       "L 21.828125 44.828125 \n",
       "Q 35.75 44.828125 43.5625 38.421875 \n",
       "Q 51.375 32.03125 51.375 20.703125 \n",
       "Q 51.375 10.15625 44.265625 4.171875 \n",
       "Q 37.15625 -1.8125 24.65625 -1.8125 \n",
       "Q 19.09375 -1.8125 12.15625 -0.296875 \n",
       "z\n",
       "\" id=\"LucidaSansUnicode-35\"/>\n",
       "      </defs>\n",
       "      <g style=\"fill:#262626;\" transform=\"translate(91.110909 246.710283)scale(0.13 -0.13)\">\n",
       "       <use xlink:href=\"#LucidaSansUnicode-32\"/>\n",
       "       <use x=\"63.232422\" xlink:href=\"#LucidaSansUnicode-35\"/>\n",
       "       <use x=\"126.464844\" xlink:href=\"#LucidaSansUnicode-30\"/>\n",
       "       <use x=\"189.697266\" xlink:href=\"#LucidaSansUnicode-30\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"xtick_3\">\n",
       "     <g id=\"line2d_5\">\n",
       "      <path clip-path=\"url(#p9bb4136610)\" d=\"M 149.824574 229.688095 \n",
       "L 149.824574 12.248095 \n",
       "\" style=\"fill:none;stroke:#ffffff;stroke-dasharray:1.5,2.475;stroke-dashoffset:0;stroke-width:1.5;\"/>\n",
       "     </g>\n",
       "     <g id=\"line2d_6\"/>\n",
       "     <g id=\"text_3\">\n",
       "      <!-- 5000 -->\n",
       "      <g style=\"fill:#262626;\" transform=\"translate(133.383636 246.710283)scale(0.13 -0.13)\">\n",
       "       <use xlink:href=\"#LucidaSansUnicode-35\"/>\n",
       "       <use x=\"63.232422\" xlink:href=\"#LucidaSansUnicode-30\"/>\n",
       "       <use x=\"126.464844\" xlink:href=\"#LucidaSansUnicode-30\"/>\n",
       "       <use x=\"189.697266\" xlink:href=\"#LucidaSansUnicode-30\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"xtick_4\">\n",
       "     <g id=\"line2d_7\">\n",
       "      <path clip-path=\"url(#p9bb4136610)\" d=\"M 192.097301 229.688095 \n",
       "L 192.097301 12.248095 \n",
       "\" style=\"fill:none;stroke:#ffffff;stroke-dasharray:1.5,2.475;stroke-dashoffset:0;stroke-width:1.5;\"/>\n",
       "     </g>\n",
       "     <g id=\"line2d_8\"/>\n",
       "     <g id=\"text_4\">\n",
       "      <!-- 7500 -->\n",
       "      <defs>\n",
       "       <path d=\"M 15.046875 0 \n",
       "Q 16.453125 8.453125 19.0625 14.59375 \n",
       "Q 21.6875 20.75 28.328125 31.5 \n",
       "L 48.046875 63.234375 \n",
       "L 10.84375 63.234375 \n",
       "L 10.84375 72.265625 \n",
       "L 57.234375 72.265625 \n",
       "L 57.234375 63.234375 \n",
       "Q 29.4375 22.265625 26.125 0 \n",
       "z\n",
       "\" id=\"LucidaSansUnicode-37\"/>\n",
       "      </defs>\n",
       "      <g style=\"fill:#262626;\" transform=\"translate(175.656364 246.710283)scale(0.13 -0.13)\">\n",
       "       <use xlink:href=\"#LucidaSansUnicode-37\"/>\n",
       "       <use x=\"63.232422\" xlink:href=\"#LucidaSansUnicode-35\"/>\n",
       "       <use x=\"126.464844\" xlink:href=\"#LucidaSansUnicode-30\"/>\n",
       "       <use x=\"189.697266\" xlink:href=\"#LucidaSansUnicode-30\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"xtick_5\">\n",
       "     <g id=\"line2d_9\">\n",
       "      <path clip-path=\"url(#p9bb4136610)\" d=\"M 234.370028 229.688095 \n",
       "L 234.370028 12.248095 \n",
       "\" style=\"fill:none;stroke:#ffffff;stroke-dasharray:1.5,2.475;stroke-dashoffset:0;stroke-width:1.5;\"/>\n",
       "     </g>\n",
       "     <g id=\"line2d_10\"/>\n",
       "     <g id=\"text_5\">\n",
       "      <!-- 10000 -->\n",
       "      <defs>\n",
       "       <path d=\"M 26.171875 0 \n",
       "L 26.171875 65.046875 \n",
       "L 14.75 65.046875 \n",
       "L 14.75 71.09375 \n",
       "L 35.84375 72.90625 \n",
       "L 35.84375 0 \n",
       "z\n",
       "\" id=\"LucidaSansUnicode-31\"/>\n",
       "      </defs>\n",
       "      <g style=\"fill:#262626;\" transform=\"translate(213.818857 246.710283)scale(0.13 -0.13)\">\n",
       "       <use xlink:href=\"#LucidaSansUnicode-31\"/>\n",
       "       <use x=\"63.232422\" xlink:href=\"#LucidaSansUnicode-30\"/>\n",
       "       <use x=\"126.464844\" xlink:href=\"#LucidaSansUnicode-30\"/>\n",
       "       <use x=\"189.697266\" xlink:href=\"#LucidaSansUnicode-30\"/>\n",
       "       <use x=\"252.929688\" xlink:href=\"#LucidaSansUnicode-30\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"xtick_6\">\n",
       "     <g id=\"line2d_11\">\n",
       "      <path clip-path=\"url(#p9bb4136610)\" d=\"M 276.642756 229.688095 \n",
       "L 276.642756 12.248095 \n",
       "\" style=\"fill:none;stroke:#ffffff;stroke-dasharray:1.5,2.475;stroke-dashoffset:0;stroke-width:1.5;\"/>\n",
       "     </g>\n",
       "     <g id=\"line2d_12\"/>\n",
       "     <g id=\"text_6\">\n",
       "      <!-- 12500 -->\n",
       "      <g style=\"fill:#262626;\" transform=\"translate(256.091584 246.710283)scale(0.13 -0.13)\">\n",
       "       <use xlink:href=\"#LucidaSansUnicode-31\"/>\n",
       "       <use x=\"63.232422\" xlink:href=\"#LucidaSansUnicode-32\"/>\n",
       "       <use x=\"126.464844\" xlink:href=\"#LucidaSansUnicode-35\"/>\n",
       "       <use x=\"189.697266\" xlink:href=\"#LucidaSansUnicode-30\"/>\n",
       "       <use x=\"252.929688\" xlink:href=\"#LucidaSansUnicode-30\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"xtick_7\">\n",
       "     <g id=\"line2d_13\">\n",
       "      <path clip-path=\"url(#p9bb4136610)\" d=\"M 318.915483 229.688095 \n",
       "L 318.915483 12.248095 \n",
       "\" style=\"fill:none;stroke:#ffffff;stroke-dasharray:1.5,2.475;stroke-dashoffset:0;stroke-width:1.5;\"/>\n",
       "     </g>\n",
       "     <g id=\"line2d_14\"/>\n",
       "     <g id=\"text_7\">\n",
       "      <!-- 15000 -->\n",
       "      <g style=\"fill:#262626;\" transform=\"translate(298.364311 246.710283)scale(0.13 -0.13)\">\n",
       "       <use xlink:href=\"#LucidaSansUnicode-31\"/>\n",
       "       <use x=\"63.232422\" xlink:href=\"#LucidaSansUnicode-35\"/>\n",
       "       <use x=\"126.464844\" xlink:href=\"#LucidaSansUnicode-30\"/>\n",
       "       <use x=\"189.697266\" xlink:href=\"#LucidaSansUnicode-30\"/>\n",
       "       <use x=\"252.929688\" xlink:href=\"#LucidaSansUnicode-30\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"xtick_8\">\n",
       "     <g id=\"line2d_15\">\n",
       "      <path clip-path=\"url(#p9bb4136610)\" d=\"M 361.18821 229.688095 \n",
       "L 361.18821 12.248095 \n",
       "\" style=\"fill:none;stroke:#ffffff;stroke-dasharray:1.5,2.475;stroke-dashoffset:0;stroke-width:1.5;\"/>\n",
       "     </g>\n",
       "     <g id=\"line2d_16\"/>\n",
       "     <g id=\"text_8\">\n",
       "      <!-- 17500 -->\n",
       "      <g style=\"fill:#262626;\" transform=\"translate(340.637038 246.710283)scale(0.13 -0.13)\">\n",
       "       <use xlink:href=\"#LucidaSansUnicode-31\"/>\n",
       "       <use x=\"63.232422\" xlink:href=\"#LucidaSansUnicode-37\"/>\n",
       "       <use x=\"126.464844\" xlink:href=\"#LucidaSansUnicode-35\"/>\n",
       "       <use x=\"189.697266\" xlink:href=\"#LucidaSansUnicode-30\"/>\n",
       "       <use x=\"252.929688\" xlink:href=\"#LucidaSansUnicode-30\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"text_9\">\n",
       "     <!-- time [sec] -->\n",
       "     <defs>\n",
       "      <path d=\"M 33.25 0 \n",
       "Q 28.8125 -1.21875 25.53125 -1.21875 \n",
       "Q 18.21875 -1.21875 14.109375 2.96875 \n",
       "Q 10.015625 7.171875 10.015625 14.65625 \n",
       "L 10.015625 45.796875 \n",
       "L 3.375 45.796875 \n",
       "L 3.375 53.03125 \n",
       "L 10.015625 53.03125 \n",
       "L 10.015625 62.640625 \n",
       "L 19.625 63.578125 \n",
       "L 19.625 53.03125 \n",
       "L 33.5 53.03125 \n",
       "L 33.5 45.796875 \n",
       "L 19.625 45.796875 \n",
       "L 19.625 16.40625 \n",
       "Q 19.625 6 28.609375 6 \n",
       "Q 30.515625 6 33.25 6.640625 \n",
       "z\n",
       "\" id=\"LucidaSansUnicode-74\"/>\n",
       "      <path d=\"M 9.625 0 \n",
       "L 9.625 53.03125 \n",
       "L 19.234375 53.03125 \n",
       "L 19.234375 0 \n",
       "z\n",
       "M 9.625 62.640625 \n",
       "L 9.625 72.265625 \n",
       "L 19.234375 72.265625 \n",
       "L 19.234375 62.640625 \n",
       "z\n",
       "\" id=\"LucidaSansUnicode-69\"/>\n",
       "      <path d=\"M 9.625 0 \n",
       "L 9.625 53.03125 \n",
       "L 19.234375 53.03125 \n",
       "L 19.234375 43.0625 \n",
       "Q 26.265625 54.203125 37.3125 54.203125 \n",
       "Q 47.953125 54.203125 51.8125 43.0625 \n",
       "Q 58.640625 54.25 69.578125 54.25 \n",
       "Q 76.5625 54.25 80.4375 50.125 \n",
       "Q 84.328125 46 84.328125 38.578125 \n",
       "L 84.328125 0 \n",
       "L 74.65625 0 \n",
       "L 74.65625 37.0625 \n",
       "Q 74.65625 46.140625 67.484375 46.140625 \n",
       "Q 60.015625 46.140625 51.8125 35.546875 \n",
       "L 51.8125 0 \n",
       "L 42.140625 0 \n",
       "L 42.140625 37.0625 \n",
       "Q 42.140625 46.1875 34.8125 46.1875 \n",
       "Q 27.546875 46.1875 19.234375 35.546875 \n",
       "L 19.234375 0 \n",
       "z\n",
       "\" id=\"LucidaSansUnicode-6d\"/>\n",
       "      <path d=\"M 48.484375 1.703125 \n",
       "Q 38.8125 -1.21875 31.9375 -1.21875 \n",
       "Q 20.171875 -1.21875 12.796875 6.5625 \n",
       "Q 5.421875 14.359375 5.421875 26.765625 \n",
       "Q 5.421875 38.8125 11.9375 46.53125 \n",
       "Q 18.453125 54.25 28.609375 54.25 \n",
       "Q 38.234375 54.25 43.484375 47.390625 \n",
       "Q 48.734375 40.53125 48.734375 27.9375 \n",
       "L 48.6875 25 \n",
       "L 15.234375 25 \n",
       "Q 17.328125 6.109375 33.734375 6.109375 \n",
       "Q 39.75 6.109375 48.484375 9.328125 \n",
       "z\n",
       "M 15.671875 32.234375 \n",
       "L 39.0625 32.234375 \n",
       "Q 39.0625 47.015625 28.03125 47.015625 \n",
       "Q 16.9375 47.015625 15.671875 32.234375 \n",
       "z\n",
       "\" id=\"LucidaSansUnicode-65\"/>\n",
       "      <path id=\"LucidaSansUnicode-20\"/>\n",
       "      <path d=\"M 9.625 -14.453125 \n",
       "L 9.625 77.09375 \n",
       "L 28.90625 77.09375 \n",
       "L 28.90625 69.875 \n",
       "L 18.0625 69.875 \n",
       "L 18.0625 -7.234375 \n",
       "L 28.90625 -7.234375 \n",
       "L 28.90625 -14.453125 \n",
       "z\n",
       "\" id=\"LucidaSansUnicode-5b\"/>\n",
       "      <path d=\"M 7.671875 1.859375 \n",
       "L 7.671875 10.6875 \n",
       "Q 17.09375 6 24.078125 6 \n",
       "Q 28.21875 6 30.953125 8.25 \n",
       "Q 33.6875 10.5 33.6875 13.875 \n",
       "Q 33.6875 18.796875 26.03125 22.015625 \n",
       "L 20.40625 24.421875 \n",
       "Q 7.953125 29.59375 7.953125 39.3125 \n",
       "Q 7.953125 46.234375 12.859375 50.21875 \n",
       "Q 17.78125 54.203125 26.3125 54.203125 \n",
       "Q 30.765625 54.203125 37.3125 52.984375 \n",
       "L 39.3125 52.59375 \n",
       "L 39.3125 44.578125 \n",
       "Q 31.25 46.96875 26.46875 46.96875 \n",
       "Q 22.265625 46.96875 19.75 45.140625 \n",
       "Q 17.234375 43.3125 17.234375 40.234375 \n",
       "Q 17.234375 35.890625 24.265625 32.90625 \n",
       "L 28.90625 30.953125 \n",
       "Q 36.765625 27.640625 40.03125 23.953125 \n",
       "Q 43.3125 20.265625 43.3125 14.75 \n",
       "Q 43.3125 7.765625 37.796875 3.265625 \n",
       "Q 32.28125 -1.21875 23.6875 -1.21875 \n",
       "Q 17.09375 -1.21875 7.671875 1.859375 \n",
       "z\n",
       "\" id=\"LucidaSansUnicode-73\"/>\n",
       "      <path d=\"M 46.4375 1.609375 \n",
       "Q 37.109375 -1.21875 30.328125 -1.21875 \n",
       "Q 19.578125 -1.21875 12.5 6.734375 \n",
       "Q 5.421875 14.703125 5.421875 26.8125 \n",
       "Q 5.421875 39.75 12.421875 46.96875 \n",
       "Q 19.4375 54.203125 31.984375 54.203125 \n",
       "Q 38.1875 54.203125 45.84375 52.484375 \n",
       "L 45.84375 44.4375 \n",
       "Q 37.703125 46.828125 32.5625 46.828125 \n",
       "Q 25.140625 46.828125 20.703125 41.28125 \n",
       "Q 16.265625 35.75 16.265625 26.515625 \n",
       "Q 16.265625 17.578125 20.84375 12.0625 \n",
       "Q 25.4375 6.546875 32.859375 6.546875 \n",
       "Q 39.453125 6.546875 46.4375 9.90625 \n",
       "z\n",
       "\" id=\"LucidaSansUnicode-63\"/>\n",
       "      <path d=\"M 22.90625 77.09375 \n",
       "L 22.90625 -14.453125 \n",
       "L 3.609375 -14.453125 \n",
       "L 3.609375 -7.234375 \n",
       "L 14.453125 -7.234375 \n",
       "L 14.453125 69.875 \n",
       "L 3.609375 69.875 \n",
       "L 3.609375 77.09375 \n",
       "z\n",
       "\" id=\"LucidaSansUnicode-5d\"/>\n",
       "     </defs>\n",
       "     <g style=\"fill:#262626;\" transform=\"translate(179.862188 265.551845)scale(0.16 -0.16)\">\n",
       "      <use xlink:href=\"#LucidaSansUnicode-74\"/>\n",
       "      <use x=\"37.402344\" xlink:href=\"#LucidaSansUnicode-69\"/>\n",
       "      <use x=\"66.308594\" xlink:href=\"#LucidaSansUnicode-6d\"/>\n",
       "      <use x=\"159.667969\" xlink:href=\"#LucidaSansUnicode-65\"/>\n",
       "      <use x=\"215.380859\" xlink:href=\"#LucidaSansUnicode-20\"/>\n",
       "      <use x=\"247.021484\" xlink:href=\"#LucidaSansUnicode-5b\"/>\n",
       "      <use x=\"279.541016\" xlink:href=\"#LucidaSansUnicode-73\"/>\n",
       "      <use x=\"330.517578\" xlink:href=\"#LucidaSansUnicode-65\"/>\n",
       "      <use x=\"386.230469\" xlink:href=\"#LucidaSansUnicode-63\"/>\n",
       "      <use x=\"437.451172\" xlink:href=\"#LucidaSansUnicode-5d\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"matplotlib.axis_2\">\n",
       "    <g id=\"ytick_1\">\n",
       "     <g id=\"line2d_17\">\n",
       "      <path clip-path=\"url(#p9bb4136610)\" d=\"M 50.060938 214.8443 \n",
       "L 384.860938 214.8443 \n",
       "\" style=\"fill:none;stroke:#ffffff;stroke-dasharray:1.5,2.475;stroke-dashoffset:0;stroke-width:1.5;\"/>\n",
       "     </g>\n",
       "     <g id=\"line2d_18\"/>\n",
       "     <g id=\"text_10\">\n",
       "      <!-- 12 -->\n",
       "      <g style=\"fill:#262626;\" transform=\"translate(26.62 219.855394)scale(0.13 -0.13)\">\n",
       "       <use xlink:href=\"#LucidaSansUnicode-31\"/>\n",
       "       <use x=\"63.232422\" xlink:href=\"#LucidaSansUnicode-32\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_2\">\n",
       "     <g id=\"line2d_19\">\n",
       "      <path clip-path=\"url(#p9bb4136610)\" d=\"M 50.060938 181.072099 \n",
       "L 384.860938 181.072099 \n",
       "\" style=\"fill:none;stroke:#ffffff;stroke-dasharray:1.5,2.475;stroke-dashoffset:0;stroke-width:1.5;\"/>\n",
       "     </g>\n",
       "     <g id=\"line2d_20\"/>\n",
       "     <g id=\"text_11\">\n",
       "      <!-- 14 -->\n",
       "      <defs>\n",
       "       <path d=\"M 37.640625 0 \n",
       "L 37.640625 20.453125 \n",
       "L 5.125 20.453125 \n",
       "L 5.125 27.734375 \n",
       "L 37.640625 72.265625 \n",
       "L 46.6875 72.265625 \n",
       "L 46.6875 28.328125 \n",
       "L 56.34375 28.328125 \n",
       "L 56.34375 20.453125 \n",
       "L 46.6875 20.453125 \n",
       "L 46.6875 0 \n",
       "z\n",
       "M 14.546875 28.328125 \n",
       "L 38.28125 28.328125 \n",
       "L 38.28125 60.453125 \n",
       "z\n",
       "\" id=\"LucidaSansUnicode-34\"/>\n",
       "      </defs>\n",
       "      <g style=\"fill:#262626;\" transform=\"translate(26.62 186.083193)scale(0.13 -0.13)\">\n",
       "       <use xlink:href=\"#LucidaSansUnicode-31\"/>\n",
       "       <use x=\"63.232422\" xlink:href=\"#LucidaSansUnicode-34\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_3\">\n",
       "     <g id=\"line2d_21\">\n",
       "      <path clip-path=\"url(#p9bb4136610)\" d=\"M 50.060938 147.299898 \n",
       "L 384.860938 147.299898 \n",
       "\" style=\"fill:none;stroke:#ffffff;stroke-dasharray:1.5,2.475;stroke-dashoffset:0;stroke-width:1.5;\"/>\n",
       "     </g>\n",
       "     <g id=\"line2d_22\"/>\n",
       "     <g id=\"text_12\">\n",
       "      <!-- 16 -->\n",
       "      <defs>\n",
       "       <path d=\"M 17.1875 37.453125 \n",
       "Q 24.078125 46.390625 34.515625 46.390625 \n",
       "Q 44 46.390625 50.03125 40.109375 \n",
       "Q 56.0625 33.84375 56.0625 24.03125 \n",
       "Q 56.0625 12.75 49.296875 5.46875 \n",
       "Q 42.53125 -1.8125 32.078125 -1.8125 \n",
       "Q 20.171875 -1.8125 13.421875 7.828125 \n",
       "Q 6.6875 17.484375 6.6875 34.515625 \n",
       "Q 6.6875 57.46875 18.453125 68.21875 \n",
       "Q 24.859375 74.125 36.140625 74.125 \n",
       "Q 42.671875 74.125 52.34375 71.390625 \n",
       "L 52.34375 62.984375 \n",
       "Q 42.140625 66.890625 35.546875 66.890625 \n",
       "Q 17.1875 66.890625 17.1875 37.453125 \n",
       "z\n",
       "M 46.390625 22.125 \n",
       "Q 46.390625 30.125 42.671875 34.78125 \n",
       "Q 38.96875 39.453125 32.515625 39.453125 \n",
       "Q 26.375 39.453125 21.96875 35.34375 \n",
       "Q 17.578125 31.25 17.578125 25.484375 \n",
       "Q 17.578125 16.40625 21.75 10.90625 \n",
       "Q 25.921875 5.421875 32.8125 5.421875 \n",
       "Q 39.0625 5.421875 42.71875 9.90625 \n",
       "Q 46.390625 14.40625 46.390625 22.125 \n",
       "z\n",
       "\" id=\"LucidaSansUnicode-36\"/>\n",
       "      </defs>\n",
       "      <g style=\"fill:#262626;\" transform=\"translate(26.62 152.310992)scale(0.13 -0.13)\">\n",
       "       <use xlink:href=\"#LucidaSansUnicode-31\"/>\n",
       "       <use x=\"63.232422\" xlink:href=\"#LucidaSansUnicode-36\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_4\">\n",
       "     <g id=\"line2d_23\">\n",
       "      <path clip-path=\"url(#p9bb4136610)\" d=\"M 50.060938 113.527697 \n",
       "L 384.860938 113.527697 \n",
       "\" style=\"fill:none;stroke:#ffffff;stroke-dasharray:1.5,2.475;stroke-dashoffset:0;stroke-width:1.5;\"/>\n",
       "     </g>\n",
       "     <g id=\"line2d_24\"/>\n",
       "     <g id=\"text_13\">\n",
       "      <!-- 18 -->\n",
       "      <defs>\n",
       "       <path d=\"M 23 39.359375 \n",
       "Q 17.046875 43.5 14.71875 47.0625 \n",
       "Q 12.40625 50.640625 12.40625 55.671875 \n",
       "Q 12.40625 63.875 18.40625 69 \n",
       "Q 24.421875 74.125 34.078125 74.125 \n",
       "Q 43.015625 74.125 48.5625 69.78125 \n",
       "Q 54.109375 65.4375 54.109375 58.5 \n",
       "Q 54.109375 48 40.625 39.359375 \n",
       "Q 56.984375 31.34375 56.984375 18.703125 \n",
       "Q 56.984375 9.671875 50.140625 3.921875 \n",
       "Q 43.3125 -1.8125 32.5625 -1.8125 \n",
       "Q 21.828125 -1.8125 15.234375 3.578125 \n",
       "Q 8.640625 8.984375 8.640625 17.78125 \n",
       "Q 8.640625 30.765625 23 39.359375 \n",
       "z\n",
       "M 35.25 42.390625 \n",
       "Q 45.359375 49.03125 45.359375 57.171875 \n",
       "Q 45.359375 61.53125 41.96875 64.1875 \n",
       "Q 38.578125 66.84375 33.109375 66.84375 \n",
       "Q 27.78125 66.84375 24.453125 64.234375 \n",
       "Q 21.140625 61.625 21.140625 57.46875 \n",
       "Q 21.140625 53.421875 24.140625 50.1875 \n",
       "Q 27.15625 46.96875 35.25 42.390625 \n",
       "z\n",
       "M 28.21875 35.453125 \n",
       "Q 22.40625 30.953125 20.328125 27.609375 \n",
       "Q 18.265625 24.265625 18.265625 19.28125 \n",
       "Q 18.265625 13.09375 22.265625 9.25 \n",
       "Q 26.265625 5.421875 32.8125 5.421875 \n",
       "Q 39.109375 5.421875 43.09375 8.6875 \n",
       "Q 47.078125 11.96875 47.078125 17.1875 \n",
       "Q 47.078125 21.390625 44.609375 24.296875 \n",
       "Q 42.140625 27.203125 35.0625 31.390625 \n",
       "z\n",
       "\" id=\"LucidaSansUnicode-38\"/>\n",
       "      </defs>\n",
       "      <g style=\"fill:#262626;\" transform=\"translate(26.62 118.538791)scale(0.13 -0.13)\">\n",
       "       <use xlink:href=\"#LucidaSansUnicode-31\"/>\n",
       "       <use x=\"63.232422\" xlink:href=\"#LucidaSansUnicode-38\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_5\">\n",
       "     <g id=\"line2d_25\">\n",
       "      <path clip-path=\"url(#p9bb4136610)\" d=\"M 50.060938 79.755496 \n",
       "L 384.860938 79.755496 \n",
       "\" style=\"fill:none;stroke:#ffffff;stroke-dasharray:1.5,2.475;stroke-dashoffset:0;stroke-width:1.5;\"/>\n",
       "     </g>\n",
       "     <g id=\"line2d_26\"/>\n",
       "     <g id=\"text_14\">\n",
       "      <!-- 20 -->\n",
       "      <g style=\"fill:#262626;\" transform=\"translate(26.62 84.76659)scale(0.13 -0.13)\">\n",
       "       <use xlink:href=\"#LucidaSansUnicode-32\"/>\n",
       "       <use x=\"63.232422\" xlink:href=\"#LucidaSansUnicode-30\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_6\">\n",
       "     <g id=\"line2d_27\">\n",
       "      <path clip-path=\"url(#p9bb4136610)\" d=\"M 50.060938 45.983295 \n",
       "L 384.860938 45.983295 \n",
       "\" style=\"fill:none;stroke:#ffffff;stroke-dasharray:1.5,2.475;stroke-dashoffset:0;stroke-width:1.5;\"/>\n",
       "     </g>\n",
       "     <g id=\"line2d_28\"/>\n",
       "     <g id=\"text_15\">\n",
       "      <!-- 22 -->\n",
       "      <g style=\"fill:#262626;\" transform=\"translate(26.62 50.994389)scale(0.13 -0.13)\">\n",
       "       <use xlink:href=\"#LucidaSansUnicode-32\"/>\n",
       "       <use x=\"63.232422\" xlink:href=\"#LucidaSansUnicode-32\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_7\">\n",
       "     <g id=\"line2d_29\">\n",
       "      <path clip-path=\"url(#p9bb4136610)\" d=\"M 50.060938 12.211094 \n",
       "L 384.860938 12.211094 \n",
       "\" style=\"fill:none;stroke:#ffffff;stroke-dasharray:1.5,2.475;stroke-dashoffset:0;stroke-width:1.5;\"/>\n",
       "     </g>\n",
       "     <g id=\"line2d_30\"/>\n",
       "     <g id=\"text_16\">\n",
       "      <!-- 24 -->\n",
       "      <g style=\"fill:#262626;\" transform=\"translate(26.62 17.222187)scale(0.13 -0.13)\">\n",
       "       <use xlink:href=\"#LucidaSansUnicode-32\"/>\n",
       "       <use x=\"63.232422\" xlink:href=\"#LucidaSansUnicode-34\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"text_17\">\n",
       "     <!-- $\\left\\langle m \\right\\rangle$ -->\n",
       "     <defs>\n",
       "      <path d=\"M 33.5 71.296875 \n",
       "L 14.296875 25.09375 \n",
       "L 33.40625 -21.296875 \n",
       "L 28.203125 -21.296875 \n",
       "L 7.703125 24.09375 \n",
       "L 7.703125 26.09375 \n",
       "L 28.296875 71.296875 \n",
       "z\n",
       "\" id=\"STIXGeneral-Regular-27e8\"/>\n",
       "      <path d=\"M 25.296875 45.90625 \n",
       "L 23.90625 40.203125 \n",
       "Q 30.296875 46.296875 39.296875 46.296875 \n",
       "Q 48.09375 46.296875 50.796875 39.796875 \n",
       "Q 58.203125 46.296875 66.59375 46.296875 \n",
       "Q 71.5 46.296875 74.5 43.6875 \n",
       "Q 77.5 41.09375 77.5 36.09375 \n",
       "Q 77.5 34 76.90625 31.5 \n",
       "L 69.09375 0 \n",
       "L 61.40625 0 \n",
       "L 68.703125 29.296875 \n",
       "Q 69.296875 31.703125 69.296875 33.5 \n",
       "Q 69.296875 39.203125 63.203125 39.203125 \n",
       "Q 60.5 39.203125 56 36.640625 \n",
       "Q 51.5 34.09375 49.203125 30.203125 \n",
       "L 41.703125 0 \n",
       "L 34 0 \n",
       "L 41.09375 29 \n",
       "Q 41.796875 32.203125 41.796875 33.5 \n",
       "Q 41.796875 39.203125 35.90625 39.203125 \n",
       "Q 33.203125 39.203125 28.640625 36.703125 \n",
       "Q 24.09375 34.203125 21.703125 30.203125 \n",
       "L 14.203125 0 \n",
       "L 6.5 0 \n",
       "L 17.40625 44 \n",
       "L 24.5 46.40625 \n",
       "z\n",
       "\" id=\"STIXGeneral-Italic-1d62e\"/>\n",
       "      <path d=\"M 32.296875 24 \n",
       "L 11.703125 -21.296875 \n",
       "L 6.5 -21.296875 \n",
       "L 25.703125 25 \n",
       "L 6.59375 71.296875 \n",
       "L 11.796875 71.296875 \n",
       "L 32.296875 26 \n",
       "z\n",
       "\" id=\"STIXGeneral-Regular-27e9\"/>\n",
       "     </defs>\n",
       "     <g style=\"fill:#262626;\" transform=\"translate(19.1 133.448095)rotate(-90)scale(0.16 -0.16)\">\n",
       "      <use transform=\"translate(0 0.703125)\" xlink:href=\"#STIXGeneral-Regular-27e8\"/>\n",
       "      <use transform=\"translate(39.999985 0.703125)\" xlink:href=\"#STIXGeneral-Italic-1d62e\"/>\n",
       "      <use transform=\"translate(115.599976 0.703125)\" xlink:href=\"#STIXGeneral-Regular-27e9\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"line2d_31\">\n",
       "    <path clip-path=\"url(#p9bb4136610)\" d=\"M 65.279119 219.804459 \n",
       "L 85.570028 219.804459 \n",
       "L 85.570028 219.804459 \n",
       "L 86.384379 173.399829 \n",
       "L 87.19873 137.888893 \n",
       "L 88.013081 110.71431 \n",
       "L 88.827432 89.919086 \n",
       "L 89.641783 74.005632 \n",
       "L 90.456134 61.827936 \n",
       "L 91.270484 52.509 \n",
       "L 92.084835 45.377723 \n",
       "L 92.899186 39.920549 \n",
       "L 93.713537 35.744472 \n",
       "L 94.527888 32.548748 \n",
       "L 95.477964 29.75556 \n",
       "L 96.42804 27.711305 \n",
       "L 97.378116 26.215174 \n",
       "L 98.328192 25.120197 \n",
       "L 99.413993 24.223413 \n",
       "L 100.635519 23.531847 \n",
       "L 101.992771 23.02801 \n",
       "L 103.757198 22.633563 \n",
       "L 106.064525 22.366704 \n",
       "L 109.593379 22.205217 \n",
       "L 117.058262 22.137764 \n",
       "L 126.151847 22.131732 \n",
       "L 126.151847 22.131732 \n",
       "L 127.03406 71.871721 \n",
       "L 127.916273 109.095739 \n",
       "L 128.798487 136.953144 \n",
       "L 129.6807 157.800852 \n",
       "L 130.562914 173.402686 \n",
       "L 131.445127 185.078667 \n",
       "L 132.327341 193.816646 \n",
       "L 133.209554 200.3559 \n",
       "L 134.091768 205.249701 \n",
       "L 134.973981 208.912079 \n",
       "L 135.856194 211.652911 \n",
       "L 136.738408 213.704072 \n",
       "L 137.620621 215.239101 \n",
       "L 138.570697 216.463206 \n",
       "L 139.588636 217.413012 \n",
       "L 140.674437 218.130561 \n",
       "L 141.963826 218.708617 \n",
       "L 143.524665 219.148262 \n",
       "L 145.49268 219.460726 \n",
       "L 146.442756 219.552892 \n",
       "L 146.442756 219.552892 \n",
       "L 147.257107 173.207319 \n",
       "L 148.071457 137.741576 \n",
       "L 148.885808 110.601576 \n",
       "L 149.700159 89.832818 \n",
       "L 150.51451 73.9396 \n",
       "L 151.328861 61.777413 \n",
       "L 152.143212 52.470344 \n",
       "L 152.957563 45.348148 \n",
       "L 153.771913 39.897913 \n",
       "L 154.586264 35.727149 \n",
       "L 155.400615 32.535489 \n",
       "L 156.350691 29.745846 \n",
       "L 157.300767 27.704193 \n",
       "L 158.250843 26.209967 \n",
       "L 159.200919 25.116388 \n",
       "L 160.28672 24.220752 \n",
       "L 161.508247 23.530069 \n",
       "L 162.865498 23.026873 \n",
       "L 164.629925 22.632923 \n",
       "L 166.937252 22.366399 \n",
       "L 170.466106 22.205119 \n",
       "L 177.930989 22.137757 \n",
       "L 187.024574 22.131732 \n",
       "L 187.024574 22.131732 \n",
       "L 187.906787 71.871721 \n",
       "L 188.789001 109.09574 \n",
       "L 189.671214 136.953144 \n",
       "L 190.553428 157.800852 \n",
       "L 191.435641 173.402686 \n",
       "L 192.317854 185.078667 \n",
       "L 193.200068 193.816644 \n",
       "L 194.082281 200.355897 \n",
       "L 194.964495 205.249693 \n",
       "L 195.846708 208.912079 \n",
       "L 196.728922 211.65291 \n",
       "L 197.611135 213.704071 \n",
       "L 198.493349 215.239097 \n",
       "L 199.443425 216.463203 \n",
       "L 200.461363 217.41301 \n",
       "L 201.547164 218.130559 \n",
       "L 202.836553 218.708615 \n",
       "L 204.397392 219.148261 \n",
       "L 206.365407 219.460727 \n",
       "L 207.315483 219.552891 \n",
       "L 207.315483 219.552891 \n",
       "L 208.129834 173.207318 \n",
       "L 208.944185 137.741575 \n",
       "L 209.758536 110.601575 \n",
       "L 210.572886 89.832818 \n",
       "L 211.387237 73.939616 \n",
       "L 212.201588 61.777421 \n",
       "L 213.015939 52.470343 \n",
       "L 213.83029 45.348138 \n",
       "L 214.644641 39.897901 \n",
       "L 215.458992 35.727143 \n",
       "L 216.273342 32.535487 \n",
       "L 217.223418 29.74585 \n",
       "L 218.173495 27.7042 \n",
       "L 219.123571 26.209976 \n",
       "L 220.073647 25.116395 \n",
       "L 221.159448 24.220754 \n",
       "L 222.380974 23.530066 \n",
       "L 223.738225 23.026868 \n",
       "L 225.502652 22.632922 \n",
       "L 227.80998 22.366406 \n",
       "L 231.338834 22.205124 \n",
       "L 238.803716 22.137757 \n",
       "L 247.897301 22.131732 \n",
       "L 247.897301 22.131732 \n",
       "L 248.779515 71.871721 \n",
       "L 249.661728 109.09574 \n",
       "L 250.543941 136.953144 \n",
       "L 251.426155 157.800852 \n",
       "L 252.308368 173.402686 \n",
       "L 253.190582 185.078667 \n",
       "L 254.072795 193.81664 \n",
       "L 254.955009 200.355903 \n",
       "L 255.837222 205.249706 \n",
       "L 256.719436 208.912093 \n",
       "L 257.601649 211.652917 \n",
       "L 258.483862 213.704073 \n",
       "L 259.366076 215.2391 \n",
       "L 260.316152 216.463207 \n",
       "L 261.33409 217.413014 \n",
       "L 262.419892 218.130563 \n",
       "L 263.70928 218.708619 \n",
       "L 265.27012 219.148264 \n",
       "L 267.238134 219.460727 \n",
       "L 268.18821 219.552891 \n",
       "L 268.18821 219.552891 \n",
       "L 269.002561 173.207319 \n",
       "L 269.816912 137.741575 \n",
       "L 270.631263 110.601575 \n",
       "L 271.445614 89.832817 \n",
       "L 272.259965 73.939618 \n",
       "L 273.074315 61.777413 \n",
       "L 273.888666 52.470343 \n",
       "L 274.703017 45.348147 \n",
       "L 275.517368 39.897912 \n",
       "L 276.331719 35.727149 \n",
       "L 277.14607 32.53549 \n",
       "L 278.096146 29.745852 \n",
       "L 279.046222 27.704202 \n",
       "L 279.996298 26.209977 \n",
       "L 280.946374 25.116394 \n",
       "L 282.032175 24.220756 \n",
       "L 283.253701 23.530067 \n",
       "L 284.610953 23.02687 \n",
       "L 286.37538 22.632924 \n",
       "L 288.682707 22.366405 \n",
       "L 292.211561 22.205123 \n",
       "L 299.676444 22.137757 \n",
       "L 308.770028 22.131732 \n",
       "L 308.770028 22.131732 \n",
       "L 309.652242 71.871721 \n",
       "L 310.534455 109.095739 \n",
       "L 311.416669 136.953144 \n",
       "L 312.298882 157.800852 \n",
       "L 313.181096 173.402686 \n",
       "L 314.063309 185.078667 \n",
       "L 314.945522 193.816645 \n",
       "L 315.827736 200.355906 \n",
       "L 316.709949 205.249704 \n",
       "L 317.592163 208.912084 \n",
       "L 318.474376 211.652912 \n",
       "L 319.35659 213.704068 \n",
       "L 320.238803 215.239096 \n",
       "L 321.188879 216.463203 \n",
       "L 322.206818 217.413008 \n",
       "L 323.292619 218.130558 \n",
       "L 324.582008 218.708615 \n",
       "L 326.142847 219.148263 \n",
       "L 328.110861 219.460729 \n",
       "L 329.060937 219.552892 \n",
       "L 329.060937 219.552892 \n",
       "L 329.875288 173.207319 \n",
       "L 330.689639 137.741576 \n",
       "L 331.50399 110.601576 \n",
       "L 332.318341 89.832816 \n",
       "L 333.132692 73.939615 \n",
       "L 333.947043 61.777416 \n",
       "L 334.761394 52.470347 \n",
       "L 335.575744 45.348145 \n",
       "L 336.390095 39.897913 \n",
       "L 337.204446 35.727148 \n",
       "L 338.018797 32.53549 \n",
       "L 338.968873 29.745854 \n",
       "L 339.918949 27.704203 \n",
       "L 340.869025 26.209977 \n",
       "L 341.819101 25.116396 \n",
       "L 342.904902 24.220757 \n",
       "L 344.126429 23.53007 \n",
       "L 345.48368 23.026874 \n",
       "L 347.248107 22.632927 \n",
       "L 349.555434 22.366406 \n",
       "L 353.084288 22.205124 \n",
       "L 360.549171 22.137757 \n",
       "L 369.642756 22.131732 \n",
       "L 369.642756 22.131732 \n",
       "\" style=\"fill:none;stroke:#0072b2;stroke-linecap:round;stroke-width:2;\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_3\">\n",
       "    <path d=\"M 50.060938 229.688095 \n",
       "L 50.060938 12.248095 \n",
       "\" style=\"fill:none;\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_4\">\n",
       "    <path d=\"M 384.860938 229.688095 \n",
       "L 384.860938 12.248095 \n",
       "\" style=\"fill:none;\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_5\">\n",
       "    <path d=\"M 50.060937 229.688095 \n",
       "L 384.860937 229.688095 \n",
       "\" style=\"fill:none;\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_6\">\n",
       "    <path d=\"M 50.060937 12.248095 \n",
       "L 384.860937 12.248095 \n",
       "\" style=\"fill:none;\"/>\n",
       "   </g>\n",
       "  </g>\n",
       " </g>\n",
       " <defs>\n",
       "  <clipPath id=\"p9bb4136610\">\n",
       "   <rect height=\"217.44\" width=\"334.8\" x=\"50.060938\" y=\"12.248095\"/>\n",
       "  </clipPath>\n",
       " </defs>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<matplotlib.figure.Figure at 0x1a1c274ac8>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.plot(test.time, test.m1_E + test.m1_P)\n",
    "plt.xlabel('time [sec]')\n",
    "plt.ylabel(r'$\\left\\langle m \\right\\rangle$')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
