\label{sec_multi_gene}
\section{Accounting for the variability in gene copy number during the cell
cycle}

When growing in rich media bacteria can double every $\approx$ 20 minutes. With
two replication forks each traveling at $\approx$ 1000 bp per second, and a
genome of $\approx$ 5 Mbp for {\it E. coli}\cite{Moran2010}, a cell would need
$\approx$ 40 minutes to replicate its genome. The apparent paradox is solved by
the fact that cells have multiple replisomes, i.e. molecular machines that
replicate the genome running in parallel. Cells can have up to 8 copies of the
genome being replicated simultaneously depending on the growth rate
\cite{Bremer1996}.

This observation implies that during the cell cycle gene copy number varies.
This variation depends on the relative position of the gene with respect to the
replication origin, having genes close to the replication origin spending more
time with multiple copies compare to genes closer to the replication termination
site. This change in gene dosage has a direct effect on the cell-to-cell
variability in gene expression \cite{Jones2014a, Peterson2015}.

\subsection{Numerical integration of moment equations}

For our specific locus ({\it galK}) and a doubling time of $\approx$ 100 min for
our experimental conditions, cells have on average 1.4 copies of the reporter
gene during the cell cycle. What this means is that cells spend 60\% of the time
having one copy of the gene and 40\% of the time with two copies. To account for
this variability in gene copy number along the cell cycle we numerically
integrate the moment equations derived in \siref{supp_moments} for a time $t =
[0, t_s]$ with an mRNA production rate $r_m$, where $t_s$ is the time point at
which the replication fork reaches our specific locus. For the remaining time
before the cell division $t = [t_s, t_d]$ that the cell spends with two
promoters, we assume that the only parameter that changes is the mRNA production
rate from $r_m$ to $2 r_m$. This simplifying assumption ignores potential
changes in protein translation rate $r_p$ or changes in the repressor copy
number that would be reflected in changes on the repressor on rate $\kron$.
\rp{Question: Let's discuss Elowitz two-color experiment.}

\subsubsection{Initial conditions for numerical integration}

In order to define the initial conditions for the numerical integration for each
of the cell cycle stages we follow a simple procedure:
\begin{enumerate}
  \item Initialize the zeroth moment, i.e. the probability of the promoter of
  being in each of the two (unregulated) or three (regulated) states at any
  value, constrained to the fact that the sum of all zeroth moments should add
  up to one.
  \item Numerically integrate the moment dynamics (\eref{eq_mom_dynamics} in the
  main text) for time longer than the length of the cell cycle in order for the
  moment values to reach steady satte using parameters corresponding to a single
  promoter. This will set the initial conditions to be used for the single
  promoter case.
  \item Integrate the moment dynamics for a time $t = [0, t_s]$ using the
  as initial conditions the last time point from step 2.
  \item Update transcription rate from $r_m$ to $2 r_m$ corresponding to the
  cell having two promoter copies.
  \item Integrate the moment dynamics for a time $t = [t_s, t_d]$ using the
  last time point of step 3 as initial conditions.
  \item As a way to represent what the cell division does to the mean mRNA and
  mean protein copy number multiply the mRNA first moment $\ee{\bb{m}}$ and the
  protein first moment $\ee{\bb{p}}$ by $1 \over 2$. In other words if at the
  end of the n$\th$ cell cycle the mean protein count is $\ee{p}_n$, the
  next cell cycle starts with a protein count $\ee{p}_{n + 1} =
  {\ee{p}_n \over 2}$.
  \item In order to set the initial value of the rest of the moments after the
  cell division integrate the moment dynamics for a time longer than a single
  cell cycle with the constraint that the zeroth moment, the mRNA and protein
  first moments remain fixed as defined by halving the values from the previous
  step.
  \item Repeat process from step 3, using as initial conditions the last time
  point of the integration in step 7.
\end{enumerate}

Step 7 in our procedure to determine initial conditions serves as a way to
determine how higher moments change after the cell divides. Only the first
moments of both mRNA and protein can simply be halved after cell division, but
it is less clear how higher moments of the distribution change as cells halve
their content. To gain intuition for why this is not a trivial step let's
consider a binomially distributed variable $X \sim \text{Bin}(N, p)$. For the
first moment, given by $\ee{X} = Np$ is easy to see that upon halving the number
of trials $N \rightarrow N/2$, the new first moment would simply be
$\ee{X}_{\text{new}} = {N \over 2} p$. But for the second moment given by
$\ee{X^2} = Np - Np^2 + N^2p^2$ it is obvious that halving the number of
attempts, i.e. setting $\ee{X^2}_{\text{new}} = {\ee{X^2} \over 2}$ is not
equivalent to the second moment of a binomial distribution with $N/2$ trials.
The right answer for this case would be to substitute every $N$ with $N/2$
obtaining $\ee{X^2}_{\text{new}} = {N \over 2}p - {N \over 2}p^2 + \left({N
\over 2}\right)^2p^2$.

What step 7 simply does is to set higher moments of the distribution to the
expected steady state value given a fixed mRNA and protein first moment. In
other words, in order to map how higher moments of the distribution change as
cells halve their content we divide the first moments of the mRNA and protein
counts in half and integrate the rest of the moment dynamics until the values
reach steady state keeping the zero and first moments fixed. This is possible
since, as shown in \siref{supp_moments}, higher moments of the distribution only
depend on lower moments. Therefore by running the higher moment dynamics while
keeping the lower moments fixed we converge to the corresponding steady state
values of the high moments.

\fref{sfig_first_mom_cycles} (adapted from \fref{fig3_cell_cycle}(B)) shows how
the first moment of both mRNA and protein changes over several cell cycles. The
mRNA quickly relaxes to the steady state corresponding to the parameters for
both a single and two promoter copies. This is expected since the parameters for
the mRNA production were determined in the first place under this assumption
(See \siref{supp_model}). We note that there is no apparent delay before
reaching steady state on the mean mRNA count after the cell divides. This is
because the mean mRNA count for the two promoters state is exactly twice the
expected mRNA count for the single promoter state (See \siref{supp_model}).
Therefore once the mean mRNA count is halved after the cell division it is
already at the steady state value for the single promoter case. On the other
hand, given that the relaxation time to steady state is determined by the
degradation rate, the mean protein count does not reach its corresponding steady
state value for either promoter copy number state. Interestingly once a couple
of cell cycles have passed the first moment has a repetitive trajectory over
cell cycles.

\begin{figure}[h!]
	\centering \includegraphics
  {../fig/moment_dynamics_numeric/first_mom_cycles.pdf}
	\caption{\textbf{First moment dynamics over cell the cell cycle.} Mean mRNA
	(upper panel) and mean protein copy number (lower panel) as the cell cycle
	progresses. The dark shaded region delimits the fraction of the cell cycle
	that cells spend with  a single copy of the promoter. The light shaded region
	delimits the fraction of the cell cycle that cells spend with two copies of
	the promoter. For a 100 min doubling time at the {\it galK} locus cells spend
	60\% of the time with one copy of the promoter and the rest with two copies.}
  \label{sfig_first_mom_cycles}
\end{figure}

Experimentally when measuring single-cell gene expression levels using
microscopy in principle cells were sampled from any time point along their
individual cell cycles. That is because our cell cultures were not synchronized
in any form, so the moments that we determined experimentally correspond to an
average over the cell cycle.  In order to compute these averages in the
following section we discuss how to account for the fact that cells are not
uniformly distributed along the cell cycle

\subsection{Exponentially distributed ages}

As mentioned in \siref{supp_param_inference}, cells in a log phase have
exponentially distributed ages along the cell cycle, having more young cells
compared to old ones. Specifically the probability of a cell of being at any
time point in the cell cycle is given by \cite{Powell1956}
\begin{equation}
  P(a) = (\ln 2) \cdot 2^{1 - a},
  \label{seq_age_prob}
\end{equation}
where $a \in [0, 1]$ is the stage of the cell cycle, with $a = 0$ being the
start of the cycle and $a = 1$ being the division.

Our numerical integration of the moment equations gave us a time evolution of
the moments as cells progress through the cell cycle. Without loss of generality
let's focus on the first mRNA moment $\ee{m(t)}$ (the same can be applied to all
other moments). In order to calculate the first moment along the entire cell
cycle we must weigh each time point by the corresponding probability that a
cell is found in such time point. This translates to computing the integral
\begin{equation}
  \ee{m}_c = \int_{\text{beginning cell cycle}}^{\text{end cell cycle}}
                       \ee{m(t)} P(t) dt,
\end{equation}
where $\ee{m}_c$ is the mean mRNA copy number averaged over the entire cell
cycle trajectory, and $P(t)$ is the probability of a cell being at a time $t$ of
its cell cycle.

If we map each time point in the cell cycle into a fraction we can use
\eref{eq_age_prob} and compute instead
\begin{equation}
  \ee{m} = \int_0^1 \ee{m(a)} P(a) da.
  \label{seq_moment_avg}
\end{equation}
What \eref{seq_moment_avg} implies is that in order to compute the first moment
(or any moment of the distribution) we must weigh each point in the moment
dynamics by the corresponding probability of a cell being at that point along
its cell cycle. So when computing a moment we take the time trajectory of a
single cell cycle as the ones shown in \fref{sfig_first_mom_cycles} and compute
the average using \eref{seq_age_prob} to weigh each time point. We perform this
integral numerically for all moments using Simpson's rule.

\subsection{Reproducing the equilibrium picture}

Given the large variability of the first moments depicted in
\fref{sfig_first_mom_cycles} it is worth considering why a simplistic
equilibrium picture has shown to be very successful in predicting the mean
expression level under diverse conditions \cite{Garcia2011c, Brewster2014,
Barnes2018, Razo-Mejia2018}. In this section we compare the simple repression
thermodynamic model with this dynamical picture of the cell cycle. But before it
is worth recapping the assumptions that go into the equilibrium model.

\subsubsection{Steady state under the thermodynamic model}

Since our equilibrium model only work for the thermodynamic limit of large
number of particles, using this theoretical framework we are only allowed to
describe the dynamics of the first moment \cite{Phillips2015}.
\rp{I challenge this. Single Brownian particle can be in equilibrium. Not clear
what you want to say or if true.}
\mrm{Now I'm doubting about my understanding of why then we have to use the
spread the butter approach to get deviations from the mean. I originally
thought that it was because the thermodynamic models work at this limit.}
Again let's only focus on the mRNA first moment $\ee{m}$. The same principles
apply if we consider the protein first moment. We can write a dynamical system
of the form
\begin{equation}
  \dt{\ee{m}} = r_m \cdot \pbound - \gm \ee{m},
\end{equation}
where as before $r_m$ and $\gm$ are the mRNA production and degradation rates
respectively, and $\pbound$ is the probability of finding the RNAP bound to the
promoter \cite{Bintu2005a}. This dynamical system is predicted to have a single
stable fixed point that we can find by computing the steady state. When we solve
for $\ee{m}$ at steady state we find
\begin{equation}
  \ee{m}_{ss} = {r_m \over \gm} \pbound.
\end{equation}

Since we assume that the only effect that the repressor has over the regulation
of the promoter is exclusion of the RNAP from binding to the promoter, we assume
that only $\pbound$ depends on the repressor copy number $R$. Therefore when
computing the fold-change in gene expression we  are left with
\begin{equation}
  \foldchange = {\ee{m (R \neq 0)}_{ss} \over \ee{m (R = 0)}_{ss}}
              = {\pbound (R \neq 0) \over \pbound (R = 0)}.
\end{equation}
And as derived in \cite{Garcia2011c} this can be simplified to
\begin{equation}
  \foldchange = \left(1 + {R \over \Nns}e^{-\beta \eR}  \right)^{-1},
  \label{seq_fold_change_thermo}
\end{equation}
where $\beta \equiv (k_BT)^{-1}$, $\eR$ is the repress-DNA binding energy, and
$\Nns$ is the number of non-specific binding sites where the repressor can bind.

To arrive to \eref{seq_fold_change_thermo} we ignore the physiological changes
that happen during the cell cycle. One of the most important being the
variability in gene copy number that we are exploring in this section. It is
therefore worth thinking about whether or not the dynamical picture exemplified
in \fref{sfig_first_mom_cycles} can be reconciled with the predictions done by
\eref{seq_fold_change_thermo} both at the mRNA and protein level.

\subsubsection{Steady state under the kinetic model}

\mrm{Here. All of this section is not necessary since we have the parameter
inference section.}
For our kinetic picture with changes over the cell cycle the mRNA (or protein)
first moment is computed as the integral defined in \eref{seq_moment_avg}.
\fref{sfig_first_mom_cycles} upper panel suggests that for the mRNA we
could approximate the integral as the sum of the steady state value of the
single promoter copy state and the steady state value of the two promoter copies
state weighted by their respective probabilities. This is because by
construction the relaxation to steady state happens at a faster time scale
compared with the length of each of the stages along the cell cycle. In other
words if $\ee{m}_i$ is the steady state value for the state with $i$ promoters,
then we can approximate the average mRNA copy number as
\begin{equation}
  \ee{m} \approx \ee{m}_1 \int_0^{f} P(a)da +
                 \ee{m}_2 \int_f^{1} P(a)da,
\end{equation}
where $f$ is the fraction of the cell cycle that cells spend with a single
promoter. Since the steady state values $\ee{m}_i$ do not depend on the stage
of the cell cycle, we were able to take them out of the integral. We can further
simplify this equation by pretending that $P(a)$ is uniform, obtaining
\begin{equation}
  \ee{m} \approx f \cdot \ee{m}_1 + (1 - f) \cdot \ee{m}_2.
  \label{eq_mRNA_mean_simplified}
\end{equation}
These simplifications will generate discrepancies between the equilibrium and
kinetic results, but as we will show bellow, these errors are within our
experimental resolution.

For both, the unregulated and the regulated promoter it can be shown that
$\ee{m}_2 = 2 \ee{m}_1$. Therefore \eref{eq_mRNA_mean_simplified} can be written
as
\begin{equation}
  \ee{m} \approx (2 - f) \ee{m}_1.
\end{equation}
Therefore the fold-change is given by
\begin{equation}
  \foldchange = {(2 - f) \ee{m}_1(R \neq 0) \over (2 - f) \ee{m}_1(R = 0)}.
\end{equation}
This equation implicitly ignores the changes in the repressor copy number as
cells progress through the cell cycle. This again is another simplifying
assumption that will generate discrepancies between the thermodynamic and the
kinetic results.

Having the dependence on the cell cycle cancel out we arrive to the same
expression as the one defined in \mrm{ref section on parameters}
\begin{equation}
  \foldchange = \left[ 1 + {\kron \over \kroff}
                \left( {\kpoff \over \kpon + \kpoff} \right)\right]^{-1}.
  \label{eq_fold_change_kinetic}
\end{equation}
Given that the functional form of \eref{eq_fold_change_kinetic} and
\eref{eq_fold_change_thermo} is the same, and we know all of the parameters
related to the RNAP kinetics from an independent dataset (See \mrm{reference
section on parameter inference}) we can constraint the ratio ${\kron \over
\kroff}$ given the parameters determined for the equilibrium model.

\fref{fig_lacI_titration} compares the predictions of both theoretical
frameworks for varying repressor copy numbers and repressor-DNA affinities. The
solid lines are directly computed from \eref{eq_fold_change_thermo}, while the
hollow triangles and the solid circles, representing the fold-change in mRNA and
protein respectively were computed by numerically integrating the moment
equations for both the two- and the three-state promoter (See
\fref{fig_first_mom_cycles} for the unregulated case) and then averaging the
time series accounting for the probability of cells being at each stage of the
cell cycle as defined in \eref{eq_moment_avg}. The systematic deviations between
both models are a consequence of the previously mentioned assumptions. But these
differences are within what we can resolve experimentally with this system
\cite{Garcia2011c, Brewster2014, Barnes2018, Razo-Mejia2018}.

For completeness \fref{fig_IPTG_titration} compares the kinetic and equilibrium
models for the extended model of \cite{Razo-Mejia2018} in which the inducer
concentration enters into the equation. The solid line is directly computed from
Eq. 5 of \cite{Razo-Mejia2018}. The hollow triangles and solid points follow the
same procedure as for \fref{fig_lacI_titration}, where the only effect that the
inducer is assume to have in the kinetics is an effective change in the number
of active repressors, affecting therefore $\kron$. Again the systematic
deviations due to our simplified assumptions fall within the experimental
resolution of the system.

\begin{figure}[h!]
	\centering \includegraphics
  {../fig/moment_dynamics_numeric/lacI_titration.pdf}
	\caption{\textbf{Comparison of the equilibrium and kinetic reressor titration
	predictions.} The equilibrium model (solid lines) and the kinetic model with
	variation over the cell cycle (solid circles and white triangles) predictions
	are compared for varying repressor copy numbers and operator binding energy.
	The equilibrium model is directly computed from \eref{eq_fold_change_thermo}
	while the kinetic model is computed by numerically integrating the moment
	equations over several cell cycles, and then averaging over the extend of the
	cell cycle as defined in \eref{eq_moment_avg}.}
  \label{fig_lacI_titration}
\end{figure}


\begin{figure}[h!]
	\centering \includegraphics
  {../fig/moment_dynamics_numeric/IPTG_titration.pdf}
	\caption{\textbf{Comparison of the equilibrium and kinetic inducer titration
	predictions.} The equilibrium model (solid lines) and the kinetic model with
	variation over the cell cycle (solid circles and white triangles) predictions
	are compared for varying repressor copy numbers and inducer concentrations.
	The equilibrium model is directly computed as Eq. 5 of reference
	\cite{Razo-Mejia2018} while the kinetic model is computed by numerically
	integrating the moment equations over several cell cycles, and then averaging
	over the extend of the cell cycle as defined in \eref{eq_moment_avg}.}
  \label{fig_IPTG_titration}
\end{figure}

\subsection{Comparison between single- and multi-promoter kinetic model}

After these calculations it is worth questioning whether the inclusion of this
extra layer of complexity is reflected on significant changes with respect to
the simpler picture of a kinetic model that ignores the gene copy number
variability during the cell cycle. To this end we systematically computed the
average moments for varying repressor copy number and repressor-DNA affinities.
We then compare these results with the moments obtained from a single-promoter
model and their corresponding parameters (See \mrm{cite section on parameter
inference}).

\fref{fig_lacI_titration} and \fref{fig_IPTG_titration} both suggest that since
the dynamic kinetic model can reproduce the results of the equilibrium model at
the first moment level it must then by definition be able to reproduce the
results of the single-promoter model at this level (See \mrm{ref FISH mcmc
section}). The interesting comparison comes with higher moments. A useful metric
to consider for gene expression variability is the noise in gene expression
$\eta$ \cite{Shahrezaei2008}. This  quantity, defined as the standard deviation
divided by the mean, is a dimensionless quantification of how much variability
there is with respect to the mean of a distribution. As we will show below this
quantity has the advantage compared to the also commonly used metric known as
the Fano factor (variance / mean) in the sense that for experimentally
determined expression levels in fluorescent arbitrary units the noise $\eta$ is
a dimensionless quantity while the Fano factor is not.

\fref{fig_noise_comparison} shows the comparison of the predicted protein noise
between  the single- (dashed lines) and the multi-promoter model (solid lines)
for different operators and repressor copy numbers. A striking difference
between both is that the single-promoter model predicts that as the mean
expression  level increases, the standard deviation grows much slower than the
mean, giving a very small noise. In comparison the multi-promoter model has a
much higher floor for the lowest value of the noise, reflecting the expected
result that the variability in gene copy number along the cell cycle should
increase the cell-to-cell variability in gene expression \cite{Peterson2015,
Jones2014a}

\begin{figure}[h!]
	\centering \includegraphics
  {../fig/moment_dynamics_numeric/noise_comparison.pdf}
	\caption{\textbf{Comparison of the predicted protein noise between a single-
	and a multi-promoter kinetic model.} Comparison of the noise $\eta$
	(standard deviation/mean) between a kinetic model that considers a single
	promoter at all times (dashed line) and the multi-promoter model developed
	in this section (solid line) for different repressor operators. (A) Operator
	O1,  $\eR = -15.3 \; k_BT$, (B) O2, $\eR = -13.9 \; k_BT$, (C) O3, $\eR =
	-9.7 \; k_BT$}
  \label{fig_noise_comparison}
\end{figure}

\subsection{Comparison with experimental data}

Having shown that the kinetic model presented in this section can reproduce the
results from the equilibrium picture at the mean level (See
\fref{fig_lacI_titration} and \fref{fig_IPTG_titration}), but changes the
prediction for the cell-to-cell variability as quantified by the noise $\eta$
(See \fref{fig_noise_comparison}), we can asses whether or not this model is
able to predict experimental measurements of the noise. For this we take the
single cell intensity measurements (See \mrm{reference microscopy section}) to
compute the noise at the protein level.

As mentioned before this metric has the advantage over the Fano factor that for
fluorescent arbitrary units (a.u.) the noise $\eta$ is a dimensionless quantity.
To see why consider that the noise is defined as
\begin{equation}
\eta \equiv \frac{\sqrt{\left\langle p^2 \right\rangle -
                        \left\langle p \right\rangle^2}}
                        {\left\langle p \right\rangle}.
    \label{eq_noise_protein}
\end{equation}
We assume that the intensity level of a cell $I$ is linearly proportional to
the absolute protein count, i.e.
\begin{equation}
I = \alpha p,
\label{eq_calibration_factor}
\end{equation}
where $\alpha$ is the proportionality constant between a.u. and protein absolute
number $p$. Substituting this definition on \eref{eq_noise_protein} gives
\begin{equation}
  \eta = {\sqrt{\ee{(\alpha I)^2} - \ee{\alpha I}^2} \over
                \ee{\alpha I}}.
\end{equation}

Since $\alpha$ is a constant it can be taken out of the average operator
$\ee{\cdot}$, obtaining
\begin{equation}
  \eta = {\sqrt{\alpha^2 \left(\ee{I^2} -
                \ee{I}^2 \right)} \over
                \alpha \ee{I}}
       = {\sqrt{\left(\ee{I^2} - \ee{I}^2 \right)} \over
                \ee{I}}.
\end{equation}

Notice that in \eref{eq_calibration_factor} the linear proportionality between
intensity and protein count has no intercept. This ignores the autofluorescence
that cells without reporter would generate. To account for this in practice to
compute the noise from experimental intensity measurements we compute
\begin{equation}
\eta = {\sqrt{\left(\ee{(I - \ee{I_\text{auto}})^2} -
                    \ee{I - \ee{I_\text{auto}}}^2 \right)} \over
                \ee{I - \ee{I_\text{auto}}}}.
\end{equation}
where $I$ is the intensity of the objective strain and $\ee{I_\text{auto}}$ is
the mean autofluorescence intensity.

\fref{fig_noise_delta} shows the comparison between theoretical predictions and
experimental measurements for the unregulated promoter. We can see that the
single-promoter model that ignores gene copy number variability greatly
underestimates the measured noise at the protein level despite the fact that it
can greatly reproduce the noise at the mRNA level (See \mrm{ref FISH MCMC
section}). On the other hand the zero parameter fit prediction from the
multi-promoter level presented in this section has a much better agreement with
the experimental values. This indicates that changes in gene copy number along
the cell cycle are important contributors to the cell-to-cell variability.

\begin{figure}[h!]
	\centering \includegraphics
  {../fig/moment_dynamics_numeric/noise_delta_microscopy.pdf}
	\caption{\textbf{Protein noise of the unregulated promoter.} Comparison of the
	experimental noise $\eta$ for different operators with the theoretical
  predictions for the single-promoter (gray dashed line) and the
  multi-promoter model (black dashed line).}
  \label{fig_noise_delta}
\end{figure}

To further test the model predictive power we compare the predictions for the
three-state regulated promoter. \fref{fig_noise_reg} shows the theoretical
predictions for the single- and the multi-promoter model for varying repressor
copy numbers and repressor-DNA binding affinities as a function of the inducer
concentration. We can see again that the zero-parameter fit multi-promoter model
has a good agreement with the experimental data, while the single-promoter model
underestimates the noise. \mrm{Here it should be strongly emphasized that the
data was absolutely not used at any point to generate the predictions. These
are 100\% parameter free predictions. A remarkable test of the validity of the
Aztec pyramid idea in my mind.}

\begin{figure}[h!]
	\centering \includegraphics
  {../fig/moment_dynamics_numeric/noise_comparison_exp.pdf}
	\caption{\textbf{Protein noise of the regulated promoter.} Comparison of the
	experimental noise $\eta$ for different operators ((A) O1,  $\eR = -15.3 \;
	k_BT$, (B) O2, $\eR = -13.9 \; k_BT$, (C) O3, $\eR = -9.7 \; k_BT$) with the
	theoretical predictions for the single-promoter (dashed lines) and the
  multi-promoter model (solid lines).}
  \label{fig_noise_reg}
\end{figure}
