\section*{Discussion}

Building on Shannon's formulation of information theory there have been
significant efforts using this theoretical framework to understand the
information processing capabilities of biological systems and the evolutionary
consequences for organisms harboring signal transduction systems
\cite{Bergstrom2004, Taylor2007a, Tkacik2008, Polani2009, Nemenman2010,
Rivoire2011}. More recently with the mechanistic dissection of molecular
signaling pathways significant progress has been made on the question of the
physical limits of cellular detection and the role that features such as
feedback loops play in this task \cite{Bialek2005, Libby2007, Tkacik2011,
Rhee2012a, Voliotis2014a}. But the field still lacks a rigorous experimental
test of these ideas with precision measurements on a system that is tractable
both experimentally and theoretically.

In this paper we take advantage of the recent progress on the quantitative
modeling of input-output functions of genetic circuits to build a minimal model
of the so-called simple repression motif. By combining a series of studies on
this circuit spanning diverse experimental methods for measuring gene expression
we infer all parameter values of our model - allowing us to generate
parameter-free predictions for processes related to information processing. The
model parameters for our kinetic formulation of the input-output function are
informed by inferences made from equilibrium assumptions. We use the fact that
if both languages describe the same system the predictions must be
self-consistent. In other words, if the equilibrium model can only make
statements about the mean mRNA and mean protein copy number, those predictions
must be equivalent to what the kinetic model has to say about these same
quantities. This condition therefore constrains the value that the kinetic rates
in the model can take. To test whether or not the kinetic picture can reproduce
the predictions made by the equilibrium model we compare the experimental and
theoretical fold-change in protein copy number for a suite of biophysical
parameters and environmental conditions. The agreement between theory and
experiment demonstrates that both languages can indeed make consistent
predictions.

The kinetic treatment of the system brings with it increasing predictive power
compared to the equilibrium picture. Under the kinetic formulation the
predictions are not limited only to the mean but to any moment of the mRNA and
protein distribution. We first tested these novel predictions by comparing the
noise in protein copy number (standard deviation / mean) with experimental data.
Since the model is able to accurately predict the noise in protein count we set
the goal to use the model predict the entire protein distribution.
