\section{Accounting for cell-cycle dependent variability in gene dosage}

As cells progress through the cell cycle the genome has to be replicated to
guarantee that each daughter cell to receives a copy of the genetic material.
This replication of the genome implies that cells spend parts of the cell cycle
with multiple copies of each gene depending on the cellular growth rate and the
relative position of the gene with respect to the replication origin
\cite{Bremer1996}. Genes closer to the replication origin spend larger fractions
of the cell cycle with multiple copies compared to genes closer to the
replication termination site. \fref{fig3_cell_cycle}(A) depicts a schematic of
this process where the replication origin ({\it oriC}) and the relevant locus
for our experimental measurements ({\it galK}) are highlighted.

Since this change in gene copy number has been shown to have an effect on
cell-to-cell variability in gene expression \cite{Jones2014a, Peterson2015}, we
now extend our minimal model to account for these changes in gene copy number
during the cell cycle.  We assume that the only difference between the
single-copy state and the two-copies states of the cell is a doubling of the
mRNA production rate $r_m$. In particular the RNAP rates $\kpon$ and $\kpoff$
and the mRNA production rate $r_m$ inferred in section \mrm{param section}
assume that cells spend a fraction $f$ of the cell cycle  with one copy of the
promoter (mRNA production rate $r_m$) and a fraction $(1-f)$ of the cell cycle
with two copies of the promoter (mRNA production rate $2 r_m$). This inference
was performed assuming that at each cell state the mRNA level immediately
reaches the steady state value for the corresponding mRNA production rate. The
steady state assumption is justified since the timescale to reach this steady
state depends only on the degradation rate $\gm$, which for the mRNA  is much
shorter ($\approx 3$ min) than the length of the cell cycle (100 min for our
experimental conditions). Appendix \mrm{ref appendix for parameter inference}
shows that by using area as a proxy for stage in the cell cycle a two-state
promoter model is able to capture the experimental data from single molecule
mRNA counts of an unregulated promoter.

Given that we assume that the protein degradation rate $\gp$ is set by the cell
division time, we cannot assume that the protein count reaches the corresponding
steady state value for each stage along the cell cycle. In other words, cells do
not spend long enough time with two copies of the promoter for the protein to
reach the steady state value corresponding to a transcription rate of $2 r_m$.
We therefore use the dynamical equations developed in \mrm{ref moments section}
to numerically integrate the moments of the distribution dynamics with the
corresponding parameters for each phase of the cell cycle.
\fref{fig3_cell_cycle}(B) shows an example corresponding to the mean mRNA level
(upper panel) and the mean protein level (lower panel) for the case of the
two-state unregulated promoter. Since we inferred the RNAP rate parameters
assuming that mRNA reaches steady state at each stage, we see that the numerical
integration of the equations are consistent with the assumption of having the
mRNA reach a stable value at each stage (See \fref{fig3_cell_cycle}(B) upper
panel). On the other hand the mean protein level does not reach a stable value
at each of the cellular stages. Nevertheless it is interesting to observe that
after a couple of cell cycles the trajectory from cycle to cycle follows a
repetitive pattern (See \fref{fig3_cell_cycle}(B) lower panel).

\begin{figure}[h!]
	\centering \includegraphics
  {./fig/main/cell_cycle_moments_v01.pdf}
	\caption{\textbf{Accounting for gene copy number variability during the cell
	cycle.} (A) Schematic of a replicating genome. As cells progress through the
	cell cycle the genome is replicated, duplicating gene copies for a fraction of
	the cell cycle. {\it oriC} indicates the replication origin, and {\it galK}
	indicates the locus at which the experimental construct was integrated. (B)
	mean mRNA (upper panel) and mean protein (lower panel) dynamics. Cells spend a
	fraction of the cell cycle with a single copy of the promoter (light brown)
	and the rest of the cell cycle with two copies (light yellow). Black arrows
	indicate time of cell division. (C) Zero parameter-fit predictions (solid
	lines) and experimental data (circles) of the fold-change (upper row) and
	noise $\nu$ (lower row) for different repressor binding sites with different
	affinities and different repressor copy numbers per cell. White dots on the
	lower row are plotted on a different scale for visual clarity. \mrm{I need to
	discuss this with you Rob. It has to do with being 100\% honest by showing all
	of the data, but not letting some outliers with known pathologies distract the
	reader from the main point.}}
  \label{fig3_cell_cycle}
\end{figure}
