\section*{Discussion}

Building on Shannon's formulation of information theory, there have been
significant efforts using this theoretical framework to understand the
information processing capabilities of biological systems, and the evolutionary
consequences for organisms harboring signal transduction systems
\cite{Bergstrom2004, Taylor2007, Tkacik2008, Polani2009, Nemenman2010,
Rivoire2011}. Recently, with the mechanistic dissection of molecular signaling
pathways significant progress has been made on the question of the physical
limits of cellular detection and the role that features such as feedback loops
play in this task \cite{Bialek2005, Libby2007, Tkacik2011, Rhee2012a,
Voliotis2014a}. But the field still lacks a rigorous experimental test of these
ideas with precision measurements on a system that is tractable both
experimentally and theoretically.

In this paper we take advantage of the recent progress on the quantitative
modeling of input-output functions of genetic circuits to build a minimal model
of the so-called simple repression motif \cite{Phillips2019}. By combining a
series of studies on this circuit spanning diverse experimental methods for
measuring gene expression under a myriad of different conditions, we infer all
parameter values of our model - allowing us to generate parameter-free
predictions for processes related to information processing. Some of the model
parameters for our kinetic formulation of the input-output function are informed
by inferences made from equilibrium models. We use the fact that if both,
kinetic and thermodynamic languages describe the same system, the predictions
must be self-consistent. In other words, if the equilibrium model can only make
statements about the mean mRNA and mean protein copy number, those predictions
must be equivalent to what the kinetic model has to say about these same
quantities. This condition therefore constrains the values that the kinetic
rates in the model can take. To test whether or not the equilibrium picture can
reproduce the predictions made by the kinetic model we compare the experimental
and theoretical fold-change in protein copy number for a suite of biophysical
parameters and environmental conditions. The agreement between theory and
experiment demonstrates that these two frameworks can indeed make consistent
predictions.

The kinetic treatment of the system brings with it increasing predictive power
compared to the equilibrium picture. Under the kinetic formulation, the
predictions are not limited only to the mean but to any moment of the mRNA and
protein distribution. We first test these novel predictions by comparing the
noise in protein copy number (standard deviation / mean) with experimental data.
Since the model is able to accurately predict the noise in protein count we
extended our analysis to infer entire protein distributions at different input
signal concentrations by using the maximum entropy principle. What this means is
that we compute moments of the protein distribution, and then use these
moments to build an approximation to the full distribution. Finally we use these
protein distributions to compute the maximum amount of information that the
genetic circuit is able to process.

By maximizing the mutual information between input signal concentration and
output protein distribution over all possible input distributions we predict
the channel capacity for a suite of biophysical parameters such as varying
repressor protein copy number and repressor-DNA binding affinity. We compare
these theoretical channel capacity predictions with experimental determinations,
finding that our minimal model is able to predict with no free parameters this
quantity. The relevance of the channel capacity comes from its interpretation as
a metric of how precise the inference that cells can make about what the state
of the environment will be, given their response.

It is important to highlight the limitations of the work presented here. As
first reported in \cite{Razo-Mejia2018}, our model fails to capture the
steepness of the fold-change induction curve for the weakest repressor binding
site (See \fref{fig3_cell_cycle}(B)). This systematic deviation for weak binding
sites remains an unresolved problem that deserves further investigation. Also
the minimal model in \fref{fig2_minimal_model}(A), despite being widely used,
is an oversimplification of the physical picture of how the transcriptional
machinery works. Since the RNAP escapes the promoter when a transcriptional
event occurs, the arrow that produces an mRNA with rate $r_m$ should also send
the promoter back to the empty state $E$ as suggested in \cite{Phillips2015}.
Nevertheless such a model with an arrow that empties the promoter after a
transcription event cannot capture the level of cell-to-cell variability in the
data. The comparison between these models requires a deeper theoretical
analysis.

The findings of this work open the opportunity to accurately test intriguing
ideas that connect Shannon's metric of how accurately a signaling system can
infer the state of the environment, with Darwinian fitness \cite{Taylor2007}.
Beautiful work along these lines has been done in the context of the
developmental program of the early {\it Drosophila} embryo \cite{Tkacik2008,
Petkova2016}. These studies demonstrated that the input-output function of the
pair-rule genes works at channel capacity, suggesting that selection has
acted on these signaling pathways, pushing them to operate at the limit of what
the physics of these systems allows. Our system differs from the early embryo in
the sense that we have a tunable circuit with variable amounts of information
processing capabilities. Furthermore, compared with the fly embryo in which the
organism tunes both the input and output distributions over evolutionary time,
we have experimental control of the distribution of inputs that the cells are
exposed to. What this means is that instead of seeing the final result of the
evolutionary process, we can set different environmental challenges, and track
over time the evolution of the population. These experiments could shed light
into the suggestive hypothesis of information bits as a metric on which
selection acts. We see this exciting direction as part of the overall effort in
quantitative biology of predicting evolution \cite{Lassig2017}.
