\subsection{Maximum Entropy approximation}\label{sec_maxent}

Having numerically computed the moments of the mRNA and protein distributions
as cells progress through the cell cycle we now proceed to make an
approximating reconstruction of the full distributions given this limited
information. As hinted in \secref{sec_moments} the maximum entropy principle,
first proposed by E.T. Jaynes in 1957 \cite{Jaynes1957}, approximates the
entire distribution by maximizing the Shannon entropy subject to constraints
given by the values of the moments of the distribution, among other quantities
\cite{Jaynes1957}. This procedure leads to a probability distribution $P_H$ of
the form (See
\siref{supp_maxent} for full derivation)
\begin{equation}
  P_H(m, p) = {1 \over \mathcal{Z}}
              \exp \left( - \sum_{(x,y)} \lambda_{(x,y)} m^x p^y \right),
  \label{eq_maxEnt_joint}
\end{equation}
where $\lambda_{(x,y)}$ is the Lagrange multiplier associated with the
constraint set by the moment $\ee{m^x p^y}$, and $\mathcal{Z}$ is a
normalization constant. The more moments $\ee{m^x p^y}$ included as
constraints, the more accurate the approximation resulting from
\eref{eq_maxEnt_joint} becomes.

The computational challenge then becomes a minimization routine in which the
values for the Lagrange multipliers $\lambda_{(x,y)}$ that are consistent with
the constraints set by the moments values $\ee{m^x p^y}$ need to be found. This
is computationally more efficient than sampling directly out of the master
equation with a stochastic algorithm (see \siref{supp_gillespie} for further
comparison between maximum entropy estimates and the Gillespie algorithm).
\siref{supp_maxent} details our implementation of a robust algorithm to find
the values of the Lagrange multipliers. \fref{fig4_maxent}(A) shows example
predicted protein distributions reconstructed using the first six moments of
the protein distribution for a suite of different biophysical parameters and
environmental inducer concentrations. As repressor-DNA binding affinity
(columns in \fref{fig4_maxent}(A)) and repressor copy number (rows in
\fref{fig4_maxent}(A)) are varied, the responses to different signals, i.e.
inducer concentrations, overlap to varying degrees. For example, the upper
right corner frame with a weak binding site ($\eR = -9.7 \; k_BT$) and a low
repressor copy number (22 repressors per cell) have virtually identical
distributions regardless of the input inducer concentration. This means that
cells with this set of parameters cannot resolve any difference in the
concentration of the signal. As the number of repressors is increased, the
degree of overlap between distributions decreases, allowing cells to better
resolve the value of the signal input. On the opposite extreme the lower left
panel shows a strong binding site ($\eR = -15.3 \; k_BT$) and a high repressor
copy number (1740 repressors per cell). This parameter combination shows
overlap between distributions since the high degree of repression skews all
distributions towards lower copy numbers, giving again little ability for the
cells to resolve the inputs. In \fref{fig4_maxent}(B) and \siref{supp_maxent}
we show the comparison of these predicted distributions with the experimental
single-cell fluorescence distributions. Given the systematic deviation of the
predicted noise as shown in \fref{fig3_cell_cycle}, the predicted distributions
in \fref{fig4_maxent}(B) underestimate the width of the experimental
distributions by a constant multiplicative factor. In the following section we
formalize the notion of how well cells can resolve different inputs from an
information theoretic perspective via the channel capacity.

\begin{figure}[h!]
	\centering \includegraphics
  {./fig/main/fig4_maxent.pdf}
	\caption{\textbf{Maximum entropy protein distributions for varying physical
	parameters.} (A) Predicted protein distributions under different inducer
	(IPTG) concentrations for different combinations of repressor-DNA
	affinities (columns) and repressor copy numbers (rows). The first six
	moments of the protein distribution used to constrain the maximum entropy
	approximation were computed by integrating \eref{eq_gral_mom} as cells
	progressed through the cell cycle as described in \secref{sec_cell_cycle}.
	(B) Theory-experiment comparison of predicted gene expression cumulative
	distribution functions. Each panel shows two example concentrations of
	inducer (colored curves) with their corresponding theoretical predictions
	(dashed lines). Distributions were normalized to the mean expression value
	of the unregulated strain in order to compare theoretical predictions in
	discrete protein counts with experimental fluorescent measurements in
	arbitrary units.}
  \label{fig4_maxent}
\end{figure}
