\section{Results}

\subsection{Minimal model of transcriptional regulation}\label{sec_model}

We begin by defining the simple repression genetic circuit to be used throughout
this work. As a tractable circuit for which we have control over the parameters
both theoretically and experimentally we chose the so-called simple  repression
motif, a common regulatory scheme among prokaryotes \cite{Rydenfelt2014}. This
circuit consists of a single promoter with an RNA-polymerase (RNAP) binding site
and a single binding site for a transcriptional repressor \cite{Garcia2011c}.
The regulation due to the repressor occurs via exclusion of the RNAP from its
binding site when the repressor is bound, decreasing the likelihood of having a
transcription event. As with many important macromolecules, we consider the
repressor to be allosteric, meaning that it can exist in two conformations, one
in which the repressor is able to bind to the specific binding site (active
state) and one in which it cannot bind the specific binding site (inactive
state). The environmental signaling occurs via passive import of an
extracellular inducer that binds the repressor, shifting the equilibrium between
the two conformations of the repressor \cite{Razo-Mejia2018}. In previous
publications we have extensively characterized the mean response of this circuit
under different conditions using equilibrium based models \cite{Phillips2019}.
In this work we build upon these models to characterize the full distribution of
gene expression with parameters such as repressor copy number and its affinity
for the DNA are systematically varied.

Given the discrete nature of molecular species copy numbers inside cells,
chemical master equations have emerged as a useful tool to model the inherent
probability distribution of these counts \cite{Sanchez2013}. In
\fref{fig2_minimal_model}(A) we show the minimal model and the necessary set of
parameters needed to predict mRNA and protein distributions. Specifically, we
assume a three-state model where the promoter can be found 1) with RNAP bound
($B$ state), 2) empty ($E$ state) and 3) with the repressor bound ($R$ state).
These three states generate a system of coupled differential equations for each
of the three state distributions $P_B(m, p; t)$, $P_E(m, p; t)$ and $P_R(m, p;
t)$, where $m$ and $p$ are the mRNA and protein count per cell, respectively and
$t$ is the time. Given the rates shown in \fref{fig2_minimal_model}(A) we define
the system of ODEs for a specific $m$ and $p$. For the RNAP bound state we have
\begin{equation}
  \begin{aligned}
    \dt{P_B(m, p)} &=
    - \overbrace{\kpoff P_B(m, p)}^{B \rightarrow E} % P -> E
    + \overbrace{\kpon P_E(m, p)}^{E \rightarrow B}\\ % E -> P
    &+ \overbrace{r_m P_B(m-1, p)}^{m-1 \rightarrow m} % m-1 -> m
    - \overbrace{r_m P_B(m, p)}^{m \rightarrow m+1}% m -> m+1
    + \overbrace{\gm (m + 1) P_B(m+1 , p)}^{m+1 \rightarrow m} % m+1 -> m
    - \overbrace{\gm m P_B(m , p)}^{m \rightarrow m-1}\\ % m -> m-1
    &+ \overbrace{r_p m P_B(m, p - 1)}^{p-1 \rightarrow p} % p-1 -> p
    - \overbrace{r_p m P_B(m, p)}^{p \rightarrow p+1} % p -> p+1
    + \overbrace{\gp (p + 1) P_B(m, p + 1)}^{p + 1 \rightarrow p} % p+1 -> p
    - \overbrace{\gp p P_B(m, p)}^{p \rightarrow p-1}. % p -> p-1
  \end{aligned}
\end{equation}
For the empty state $E$ we have
\begin{equation}
  \begin{aligned}
    \dt{P_E(m, p)} &=
    \overbrace{\kpoff P_B(m, p)}^{B \rightarrow E} % P -> E
    - \overbrace{\kpon P_E(m, p)}^{E \rightarrow B} % E -> P
    + \overbrace{\kroff P_R(m, p)}^{R \rightarrow E} % R -> E
    - \overbrace{\kron P_E(m, p)}^{E \rightarrow R}\\ % E -> R
    &+ \overbrace{\gm (m + 1) P_E(m+1 , p)}^{m+1 \rightarrow m} % m+1 -> m
    - \overbrace{\gm m P_E(m , p)}^{m \rightarrow m-1}\\ % m -> m-1
    &+ \overbrace{r_p m P_E(m, p - 1)}^{p-1 \rightarrow p} % p-1 -> p
    - \overbrace{r_p m P_E(m, p)}^{p \rightarrow p+1} % p -> p+1
    + \overbrace{\gp (p + 1) P_E(m, p + 1)}^{p + 1 \rightarrow p} % p+1 -> p
    - \overbrace{\gp p P_E(m, p)}^{p \rightarrow p-1}. % p -> p-1
  \end{aligned}
\end{equation}
And finally, for the repressor bound state $R$ we have
\begin{equation}
  \begin{aligned}
    \dt{P_R(m, p)} &=
    - \overbrace{\kroff P_R(m, p)}^{R \rightarrow E} % R -> E
    + \overbrace{\kron P_E(m, p)}^{E \rightarrow R}\\ % E -> R
    &+ \overbrace{\gm (m + 1) P_R(m+1 , p)}^{m+1 \rightarrow m} % m+1 -> m
    - \overbrace{\gm m P_R(m , p)}^{m \rightarrow m-1}\\ % m -> m-1
    &+ \overbrace{r_p m P_R(m, p - 1)}^{p-1 \rightarrow p} % p-1 -> p
    - \overbrace{r_p m P_R(m, p)}^{p \rightarrow p+1} % p -> p+1
    + \overbrace{\gp (p + 1) P_R(m, p + 1)}^{p + 1 \rightarrow p} % p+1 -> p
    - \overbrace{\gp p P_R(m, p)}^{p \rightarrow p-1}. % p -> p-1
  \end{aligned}
\end{equation}
As we will discuss later in \secref{sec_cell_cycle} the protein degradation term
$\gp$ is set to zero since we do not consider protein degradation as a Poission
process, but rather we explicitly implement binomial partitioning as the cells
grow and divide.

It is convenient to rewrite these equations in a compact matrix notation
\cite{Sanchez2013}. For this we define the vector $\PP(m, p)$ as
\begin{equation}
  \PP(m, p) = (P_B(m, p), P_E(m, p), P_R(m, p))^T,
\end{equation}
where $T$ is the transpose. By defining the matrices $\Km$ to contain the
promoter state transitions, $\Rm$ and $\Gm$ to contain the mRNA production and
degradation terms, respectively, and $\Rp$ and $\Gp$ to contain the protein
production and degradation terms, respectively, the system of ODEs can then be
written as (See \siref{supp_model} for full definition of these matrices)
\begin{equation}
  \begin{aligned}
    \dt{\PP(m, p)} &= \left( \Km -\Rm -m\Gm -m\Rp -p\Gp \right) \PP(m, p)\\
    &+ \Rm \PP(m-1, p)
    + (m + 1) \Gm \PP(m + 1, p)\\
    &+ m \Rp \PP(m, p - 1)
    + (p + 1) \Gp \PP(m, p + 1).
  \end{aligned}
  \label{eq_cme_matrix}
\end{equation}

\subsection{Inferring parameters from published data sets}
\label{sec_param}

A decade of research in our group has characterized the simple repression motif
with an ever expanding array of predictions and corresponding experiments to
uncover the physics of this genetic circuit \cite{Phillips2019}. In doing so
we have come to understand the mean response of a single promoter in the
presence of varying levels of repressor copy numbers and repressor-DNA
affinities \cite{Garcia2011c}, due to the effect that competing binding sites
and multiple promoter copies impose \cite{Brewster2014}, and in recent work,
assisted by the Monod-Wyman-Changeux (MWC) model, we expanded the scope to the
allosteric nature of the repressor \cite{Razo-Mejia2018}. All of these studies
have exploited the simplicity and predictive power of equilibrium approximations
to these non-equilibrium systems \cite{Buchler2003}. We have also used a similar
kinetic model to the one depicted in \fref{fig2_minimal_model}(A) to study the
noise in mRNA copy number \cite{Jones2014a}. As a test case of the depth of our
theoretical understanding of the so-called ``hydrogen atom'' of transcriptional
regulation we combine all of the studies mentioned above to inform the parameter
values of the model presented in \fref{fig2_minimal_model}(A).
\fref{fig2_minimal_model}(B) schematizes the data sets and experimental
techniques used to measure gene expression along with the parameters that can be
inferred from them.

\siref{supp_param_inference} expands on the details of how the inference was
performed for each of the parameters. Briefly the RNAP rates $\kpon$ and
$\kpoff$, as well as the transcription rate $r_m$ were obtained in units of the
mRNA degradation rate $\gm$ by fitting a two-state promoter model (no state $R$
from \fref{fig2_minimal_model}(A)) \cite{Peccoud1995} to mRNA FISH data of an
unregulated promoter (no  repressor present in the cell) \cite{Jones2014a}. The
repressor on rate is assumed to be of the form $\kron = k_o [R]$ where $k_o$ is
a diffusion-limited on rate and $[R]$ is the concentration of active repressor
in the cell \cite{Jones2014a}. This concentration of active repressor is at the
same time determined by the mean repressor copy number in the cell, and the
fraction of repressors in the active state. Existing estimates of the transition
rates between conformations of allosteric molecules set them at the microsecond
scale \cite{Cui2008}. By considering this to be representative for our repressor
of interest, the separation of time-scales between the rapid conformational
changes of the repressor and the slower downstream processes such as the
open-complex formation processes allow us to model the probability of the
repressor being in the active state as an equilibrium MWC process. The
parameters of the MWC model $K_A$, $K_I$ and $\eAI$ were previously
characterized from video-microscopy and flow-cytometry data
\cite{Razo-Mejia2018}. For the repressor off rate $\kroff$ we take advantage of
the fact that the mean mRNA copy number as derived from the model in
\fref{fig2_minimal_model}(A) cast in the language of rates is of the same
functional form as the equilibrium model cast in the language of binding
energies \cite{Phillips2015}. Therefore the value of the repressor-DNA binding
energy $\eR$ constrains the value of the repressor off rate $\kroff$. These
constraints on the rates allow us to make self-consistent predictions under
both, the equilibrium and the kinetic framework.

\begin{figure}[h!]
	\centering \includegraphics
  {./fig/main/parameter_inference_v03.pdf}
	\caption{\textbf{Minimal kinetic model of transcriptional regulation for a
	simple repression architecture.} (A) Three-state promoter stochastic model of
	transcriptional regulation by a repressor. The regulation by the repressor
	occurs via exclusion of the RNAP, not allowing a promoter state in which both,
	the repressor and the RNAP are bound simultaneously. All parameters
	highlighted with colored boxes were determined with published datasets based
	on the same genetic circuit. (B) Data sets used to infer the parameter values.
	From left to right Garcia \& Phillips \cite{Garcia2011c} is used to determine
	$\kroff$ and $\kron$, Brewster et al. \cite{Brewster2014} is used to determine
	$\eAI$ and $\kron$, Razo-Mejia et al. \cite{Razo-Mejia2018} is used to
	determine $K_A$, $K_I$, and $\kron$ and Jones et al. is used to determine
	$r_m$, $\kpon$, and $\kpoff$.}
  \label{fig2_minimal_model}
\end{figure}
