---
layout: post
title: Analysis
cover: analysis.png
categories: posts
permalink: analysis
---

# Analysis files

This page contains Jupyter notebooks for each major analysis section in the
publication. You can either read our analysis rendered in HTML or download the
`.ipynb` file and run the analysis on your own machine. All necessary data files
can be accessed from the
[data](https://mrazomej.github.io/chann_cap/data)
page.

# Theory

In particular these notebooks are related to the minimal model of gene
expression and its connection to the information processing capacity.

[Unregulated promoter parameter inference](src/theory/html/chemical_master_mRNA_FISH_mcmc.html) \| Bayesian inference of the promoter rates
based on single-molecule mRNA FISH data. [\[data\]](../data/mRNA_FISH/Jones_Brewster_2014.csv)



<!-- * [Supplemental Information B - Unsupervised Gating](../../../../code/notebooks/SI_B_unsupervised_gating.html) \| An example of how our automated gating function works on raw flow cytometry data. [\[`.ipynb` file\]](../../../../code/notebooks/SI_B_unsupervised_gating.ipynb)

* [Supplemental Information C - Single-Cell Microscopy](../../../../code/notebooks/SI_C_microscopy.html) \| An example of how microscopy images were processed including single-cell segmentation and correction for non-uniform illumination. [\[`.ipynb` file\]](../../../../code/notebooks/SI_C_microscopy.ipynb)

* [Supplemental Information E - Bayesian Parameter Estimation](../../../../code/notebooks/SI_E_bayesian_parameter_estimation.html) \| A notebook covering the procedure of estimating parameter values using a Bayesian definition of probability. [\[`.ipynb` file\]](../../../../code/notebooks/SI_E_bayesian_parameter_estimation.ipynb)

* [Supplemental Information E - Global Parameter Estimation](../../../../code/notebooks/SI_E_global_fit.html) \| An annotated and thorough description of our parameter estimation considering all known
sources of error for the model parameters.[\[`.ipynb` file\]](../../../../code/notebooks/SI_E_global_fit.ipynb)

* [Supplemental Information E - Nonlinear Regression](../../../../code/notebooks/SI_E_nonlinear_regression.html) \| An example of how nonlinear regression was used to generate the starting positions of the MCMC walkers, as was used in the [global parameter estimation notebook](../../../../code/notebooks/SI_E_global_fits.html). [\[`.ipynb` file\]](../../../../code/notebooks/SI_E_nonlinear_regression.ipynb)

* [Supplemental Information H - Sensitivity Analysis](../../../../code/notebooks/SI_H_sensitivity.html) \| An explanation and analysis of how changes in the inducer dissociation constants values influence the credible region of the predicted fold-change in gene expression. [\[`.ipynb` file\]]


## Example Scripts

Our experimental pipeline included two Python scripts used for processing the
raw flow cytometry data and generating quality control plots. These scripts generated files that were saved to disk after each run.

* [Example `example_processing.py` file](../../../../code/examples/example_processing.py) \| This script reads in flow cytometry measurements from a `.csv` file, employs an unsupervised gating algorithm, computes the experimental fold-change in gene expression, and saves two `.csv` files with said fold-change values.

* [Example `analysis.py` file](../../../../code/examples/example_analysis.py) \| This script reads in the fold-change `.csv` file generated by [`example_processing.py`](../../../../code/examples/example_processing.py) and generates a plot of the intensity as a function of IPTG concentration as well as a titration of fold-change in the absence of inducer as a function of repressor copy number. -->
