---
layout: post
title: Analysis
cover: analysis.png
categories: posts
permalink: analysis
---

# Analysis files

This page contains Jupyter notebooks for each major analysis section in the
publication. You can either read our analysis rendered in HTML or download the
`.ipynb` file and run the analysis on your own machine. All necessary data files
can be accessed from the
[data](https://mrazomej.github.io/chann_cap/data)
page.

## Theory

In particular these notebooks are related to the minimal model of gene
expression and its connection to the information processing capacity.

* [Unregulated promoter parameter inference](src/theory/html/chemical_master_mRNA_FISH_mcmc.html) \| Bayesian inference of the promoter rates
based on single-molecule mRNA FISH data. [\[`data`\]](data/mRNA_FISH/Jones_Brewster_2014.csv) [\[`.ipynb` file\]](src/theory/chemical_master_mRNA_FISH_mcmc.ipynb)

* [Single promoter steady-state moments](src/theory/html/chemical_master_steady_state_moments_general.html) \| Analytical computation of the steady-state mRNA and protein moments of the distribution for the single promoter model with protein degradation as a Poisson process. [\[`.ipynb` file\]](src/theory/chemical_master_steady_state_moments_general.ipynb)

* [Moment dynamics](src/theory/html/moment_dynamics_system.html) \| Analytical derivation of the dynamical equations for the mRNA and protein moments of the distribution. [\[`.ipynb` file\]](src/theory/moment_dynamics_system.ipynb)

* [Multi promoter 'dynamical' steady-state moments](src/theory/html/moment_dynamics_cell_division.html) \| Numerical integration of the mRNA and protein moments of the distribution during the cell cycle. [\[`.ipynb` file\]](src/theory/moment_dynamics_cell_division.ipynb)

* [Maximum Entropy distributions](src/theory/html/MaxEnt_approx_joint.html) \| Inference of the maximum entropy mRNA and protein distributions. [\[`.ipynb` file\]](src/theory/MaxEnt_approx_joint.ipynb)

* [Channel capacity](src/theory/html/blahut_algorithm_channel_capacity.html) \| Numerical computation of the channel capacity. [\[`.ipynb` file\]](src/theory/blahut_algorithm_channel_capacity.ipynb)

## Data processing

* [Image analysis pipeline](src/image_analysis/image_analysis_pipeline.html) \| Example analysis of microscopy images to extract single-cell intensity values. [\[`.ipynb` file\]](src/image_analysis/image_analysis_pipeline.ipynb)

<!-- ## Example Scripts

Our experimental pipeline included two Python scripts used for processing the
raw flow cytometry data and generating quality control plots. These scripts generated files that were saved to disk after each run.

* [Example `example_processing.py` file](../../../../code/examples/example_processing.py) \| This script reads in flow cytometry measurements from a `.csv` file, employs an unsupervised gating algorithm, computes the experimental fold-change in gene expression, and saves two `.csv` files with said fold-change values.

* [Example `analysis.py` file](../../../../code/examples/example_analysis.py) \| This script reads in the fold-change `.csv` file generated by [`example_processing.py`](../../../../code/examples/example_processing.py) and generates a plot of the intensity as a function of IPTG concentration as well as a titration of fold-change in the absence of inducer as a function of repressor copy number. --> -->
